{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome Mining Script\n",
    "##### Original Author: Chun Yin Larry So | Python Interpretation: Nathan Alam\n",
    "[Github repository](https://github.com/nathanalam/genomemining)\n",
    "\n",
    "The purpose of this notebook is to provide scripts for reading bacterial genomes in search of regular expressions which seem to match a String corresponding to a candidate coding for a Lasso peptide.\n",
    "\n",
    "## Links to relevant publications:\n",
    "- [Genome mining for lasso peptides: past, present, and future](https://link.springer.com/article/10.1007/s10295-019-02197-z)\n",
    "- [Prospecting Genomes for Lasso Peptides](https://www.ncbi.nlm.nih.gov/pubmed/24142336)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comments from Perl Script\n",
    "- In this version each time MAST needs to be run on a putative maturation enzyme in a particular genome, a lookup in the database is performed to check if this protein has been analyzed by MAST before and if it has then use the values from before\n",
    "- In this version traseq will be used instead of getorf for finding precursors\n",
    "- getorf is still used for finding neighbors\n",
    "- Like v4 except that multiple proteins of the same sequence don't cause wrong locations of maturation enzymes to be reported\n",
    "- Another change is that the pattern needs to be adjusted to [MVL] instead of ^ in the beginning\n",
    "- This version fixes the problem of having a useless %AME hash and also of erasing sequences from %AME_scores\n",
    "- Precursor pattern is output into the log file\n",
    "- ORF searching behavior is changed from stop-to-stop to [MVL]-to-stop\n",
    "- Clusters of precursors are saved in clusters.txt\n",
    "- **Warning**: transeq doesn't label the -1, -2, -3 frames sequentially. Sometimes it is -1, -3, -2, sometimes some other combination. This means the precursor locations on the reverse strand are off by one sometimes\n",
    "- Take note that rank_hits expects 4 motifs for the B enzyme and 3 motifs for the C enzyme. Adjust accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import sys\n",
    "import os\n",
    "import json\n",
    "import requests\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### THE Pattern\n",
    "\n",
    "This is the pattern that we're using to identify lasso proteins. TODO - Use Machine Learning to adjust the pattern to maximize the number of valid lasso proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATTERN = 'M[A-Z]{15,45}T[A-Z][A-Z]{6,8}[DE][A-Z]{5,30}\\*'\n",
    "# PATTERN = 'CC.CGCCC...TGGC.'\n",
    "# PATTERN = '.*'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTA Function\n",
    "\n",
    "Define a function that takes as input the relative path of a FASTA formatted text file, return an object that contains a list of sequence objects. Each sequence object has a description field [\"description\"] and a sequence field [\"sequence\"].\n",
    "\n",
    "From http://www.csbio.sjtu.edu.cn/bioinf/virus-multi/example.htm, specification of a FASTA formatted file:\n",
    "- The first line of each query protein input format must begin with a greater-than (\">\") symbol in the first column. The word following the \">\" symbol is the identifier and description of the sequence, but both are optional.\n",
    "- The sequence (in single-character code) begins in a different line and ends if another line starting with a \">\" appears, which indicates the start of another query protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFASTA(name, cleanspace = 0):\n",
    "    descriptions = []\n",
    "    sequences = []\n",
    "    sequenceList = []\n",
    "    tempSequences = []     \n",
    "        \n",
    "    with open(name) as file:\n",
    "        count = -1\n",
    "        for line in file:\n",
    "            \n",
    "            if(line[0] == '>'):\n",
    "                # if begins with a >, then a description\n",
    "                descriptions.append(line[1:].replace('\\n', ''))\n",
    "                count += 1\n",
    "                # skip the first time\n",
    "                if count > 0 :\n",
    "                    # combine the tempSequences into a single string and\n",
    "                    # add it to sequences\n",
    "                    newSequence = ' '.join(tempSequences)\n",
    "                    # now remove all of the whitespaces\n",
    "                    newSequence = newSequence.replace(' ', '')\n",
    "                    newSequence = newSequence.replace('\\n', '')\n",
    "                    \n",
    "                    sequences.append(newSequence)\n",
    "                    # refresh the tempSequence list\n",
    "                    tempSequences = []\n",
    "                    \n",
    "                    sequenceList.append({\n",
    "                        \"description\": descriptions[count - 1],\n",
    "                        \"sequence\": sequences[count - 1]\n",
    "                    })\n",
    "            else:\n",
    "                tempSequences.append(line)\n",
    "                \n",
    "        # combine the tempSequences into a single string and\n",
    "        # add it to sequences\n",
    "        newSequence = ' '.join(tempSequences)\n",
    "        # now remove all of the whitespaces\n",
    "        newSequence = newSequence.replace(' ', '')\n",
    "        newSequence = newSequence.replace('\\n', '')\n",
    "\n",
    "        sequences.append(newSequence)\n",
    "        # refresh the tempSequence list\n",
    "        tempSequences = []\n",
    "        \n",
    "        sequenceList.append({\n",
    "            \"description\": descriptions[count],\n",
    "            \"sequence\": sequences[count]\n",
    "        })\n",
    "                \n",
    "                \n",
    "    if len(descriptions) != len(sequences):\n",
    "        print(\"ERROR: Number of descriptions does not match number of sequences\")\n",
    "        print(\"Number of descriptions: \" + str(len(descriptions)))\n",
    "        print(\"Number of sequences: \" + str(len(sequences)))\n",
    "        sys.exit(1);\n",
    "        \n",
    "    print(\"Read \" + str(count + 1) + \" objects from FASTA file \" + name)\n",
    "        \n",
    "    return sequenceList\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain Amino acid sequence directories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Begin by getting a list of all of the genomes available in the genomes folder alongside this script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDIRNAMES = []\n",
    "for dirname in os.listdir(\"genomes\"):\n",
    "    ## if a regular file, just add to directory\n",
    "    if (dirname.find(\".\") != -1):\n",
    "        ALLDIRNAMES.append(\"genomes/\" + dirname)\n",
    "    else:\n",
    "        for filename in os.listdir(\"genomes/\" + dirname):\n",
    "            ALLDIRNAMES.append(\"genomes/\" + dirname + \"/\" + filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831789.1.faa', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831788.1.fna', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831789.1.fna', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831790.1.faa', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831790.1.fna', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831788.1.faa', 'genomes/Streptomyces_lividans_TK24_GCA_000739105_1\\uf028/NZ_CP009124.1.fna', 'genomes/Streptomyces_lividans_TK24_GCA_000739105_1\\uf028/NZ_CP009124.1.faa', 'genomes/Streptomyces_albus_GCA_000827005_1\\uf028/NZ_CP010519.1.faa', 'genomes/Streptomyces_albus_GCA_000827005_1\\uf028/NZ_CP010519.1.fna', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020816.1.faa', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020815.1.fna', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020817.1.fna', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020816.1.fna', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020815.1.faa', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020817.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014816.1.fna', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014818.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014818.1.fna', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014819.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014817.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014819.1.fna', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014817.1.fna', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014816.1.faa', 'genomes/Frankia_CcI3_GCA_000013345_1\\uf028/NC_007777.1.fna', 'genomes/Frankia_CcI3_GCA_000013345_1\\uf028/NC_007777.1.faa', 'genomes/Streptococcus_suis_ST3_GCA_000204625_1\\uf028/NC_015433.1.fna', 'genomes/Streptococcus_suis_ST3_GCA_000204625_1\\uf028/NC_015433.1.faa', 'genomes/Streptomyces_albus_GCA_001577385_1\\uf028/NZ_CP014485.1.faa', 'genomes/Streptomyces_albus_GCA_001577385_1\\uf028/NZ_CP014485.1.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000071.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000022.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000092.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000042.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000060.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000079.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000062.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000051.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000067.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000057.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000034.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000034.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000067.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000090.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000041.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000040.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000072.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000039.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000083.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000094.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000026.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000014.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000017.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000032.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000013.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000003.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000026.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000004.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000079.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000080.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000008.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000052.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000023.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000072.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000031.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000065.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000085.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000053.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000048.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000076.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000007.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000013.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000048.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000012.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000081.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000054.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000065.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000046.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000100.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000095.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000058.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000093.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000097.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000022.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000047.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000078.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000005.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000086.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000044.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000082.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000031.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000080.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000040.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000070.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000039.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000023.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000081.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000028.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000060.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000049.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000007.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000077.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000044.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000084.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000085.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/.DS_Store', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000021.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000033.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000025.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000049.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000035.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000073.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000058.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000059.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000036.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000064.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000042.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000032.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000016.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000033.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000008.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000003.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000068.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000055.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000028.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000089.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000019.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000075.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000016.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000096.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000099.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000078.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000099.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000074.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000068.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000010.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000041.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000066.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000050.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000014.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000037.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000018.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000012.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000096.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000045.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000052.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000053.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000091.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000094.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000098.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000063.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000024.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000090.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000084.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000027.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000092.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000083.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000093.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000059.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000004.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000061.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000055.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000025.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000020.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000047.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000086.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000010.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000019.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000005.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000064.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000015.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000070.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000054.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000097.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000075.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000091.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000095.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000015.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000024.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000088.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000066.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000098.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000011.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000030.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000020.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000038.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000002.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000009.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000006.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000056.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000043.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000088.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000036.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000050.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000030.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000087.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000006.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000029.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000069.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000076.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000051.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000087.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000035.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000037.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000029.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000018.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000100.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000056.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000063.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000089.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000062.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000057.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000061.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000074.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000021.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000038.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000009.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000002.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000001.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000001.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000017.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000073.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000082.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000045.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000069.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000043.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000046.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000027.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000071.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000011.fna', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000077.fna', 'genomes/Streptomyces_albus_J1074_GCA_000359525_1\\uf028/NC_020990.1.faa', 'genomes/Streptomyces_albus_J1074_GCA_000359525_1\\uf028/NC_020990.1.fna']\n"
     ]
    }
   ],
   "source": [
    "print(ALLDIRNAMES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clear Directory\n",
    "Delete all translated directories for further testing. **ONLY DO THIS IF YOU KNOW WHAT YOU'RE DOING**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for dirname in ALLDIRNAMES:\n",
    "#     if (dirname[len(dirname) - 3:] == \"faa\"):\n",
    "#         os.remove(dirname)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the fna files, we need to convert them to amino acid sequences (or faa files). We do this using [Emboss Transeq](https://www.ebi.ac.uk/seqdb/confluence/display/JDSAT/EMBOSS+transeq+Help+and+Documentation#EMBOSStranseqHelpandDocumentation-Reference), and save the output amino acid sequences into this genome file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq, reverse_complement, translate\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "# takes a three letter codon, and returns the corresponding amino acid or 'X' if unknown\n",
    "def translate_codon(codonString):\n",
    "    if(not len(codonString) == 3):\n",
    "        raise InputError()\n",
    "        \n",
    "    try:\n",
    "        return translate(codonString, to_stop=False, table = 11)\n",
    "    except:\n",
    "        return 'X'\n",
    "\n",
    "## An adapter function for the biopython's translate, takes in a DNA sequence and returns a list of protein sequences\n",
    "def get_orfs(DNAseq):\n",
    "    AAList = []\n",
    "    \n",
    "    codonArr = []\n",
    "    seqLen = len(DNAseq) - (len(DNAseq) % 3)\n",
    "    seq = ''\n",
    "    try:\n",
    "        seq = translate(DNAseq[0:seqLen])\n",
    "    except:\n",
    "        for i in range(0, seqLen, 3):\n",
    "            codonArr.append(translate_codon(DNAseq[i:i + 3]))\n",
    "        seq = ''.join(codonArr)\n",
    "    AAList.append({\n",
    "        \"ORF\": 1,\n",
    "        \"sequence\": seq\n",
    "    })\n",
    "    \n",
    "    codonArr = []\n",
    "    seqLen = len(DNAseq) - ((len(DNAseq) - 1) % 3)\n",
    "    seq = ''\n",
    "    try:\n",
    "        seq = translate(DNAseq[1:seqLen])\n",
    "    except:\n",
    "        for i in range(1, seqLen, 3):\n",
    "            codonArr.append(translate_codon(DNAseq[i:i + 3]))\n",
    "\n",
    "        seq = ''.join(codonArr)\n",
    "    AAList.append({\n",
    "        \"ORF\": 2,\n",
    "        \"sequence\": seq\n",
    "    })\n",
    "    \n",
    "    codonArr = []\n",
    "    seqLen = len(DNAseq) - ((len(DNAseq) - 2) % 3)\n",
    "    seq = ''\n",
    "    try:\n",
    "        seq = translate(DNAseq[2:seqLen])\n",
    "    except:\n",
    "        for i in range(2, seqLen, 3):\n",
    "            codonArr.append(translate_codon(DNAseq[i:i + 3]))\n",
    "\n",
    "        seq = ''.join(codonArr)\n",
    "    AAList.append({\n",
    "        \"ORF\": 3,\n",
    "        \"sequence\": seq\n",
    "    })\n",
    "    \n",
    "    backwards_dna = reverse_complement(DNAseq)\n",
    "    codonArr = []\n",
    "    seqLen = len(backwards_dna) - (len(backwards_dna) % 3)\n",
    "    seq = ''\n",
    "    try:\n",
    "        seq = translate(backwards_dna[0:seqLen])\n",
    "    except:\n",
    "        for i in range(0, seqLen, 3):\n",
    "            codonArr.append(translate_codon(backwards_dna[i:i + 3]))\n",
    "        seq = ''.join(codonArr)\n",
    "    AAList.append({\n",
    "        \"ORF\": -1,\n",
    "        \"sequence\": seq\n",
    "    })\n",
    "    \n",
    "    codonArr = []\n",
    "    seqLen = len(backwards_dna) - ((len(backwards_dna) - 1) % 3)\n",
    "    seq = ''\n",
    "    try:\n",
    "        seq = translate(backwards_dna[1:seqLen])\n",
    "    except:\n",
    "        for i in range(1, seqLen, 3):\n",
    "            codonArr.append(translate_codon(backwards_dna[i:i + 3]))\n",
    "\n",
    "        seq = ''.join(codonArr)\n",
    "    AAList.append({\n",
    "        \"ORF\": -2,\n",
    "        \"sequence\": seq\n",
    "    })\n",
    "    \n",
    "    codonArr = []\n",
    "    seqLen = len(backwards_dna) - ((len(backwards_dna) - 2) % 3)\n",
    "    seq = ''\n",
    "    try:\n",
    "        seq = translate(backwards_dna[2:seqLen])\n",
    "    except:\n",
    "        for i in range(2, seqLen, 3):\n",
    "            codonArr.append(translate_codon(backwards_dna[i:i + 3]))\n",
    "\n",
    "        seq = ''.join(codonArr)\n",
    "    AAList.append({\n",
    "        \"ORF\": -3,\n",
    "        \"sequence\": seq\n",
    "    })\n",
    "    \n",
    "    return AAList   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'ORF': 1, 'sequence': 'MHGS'},\n",
       " {'ORF': 2, 'sequence': 'CMAA'},\n",
       " {'ORF': 3, 'sequence': 'AWQ'},\n",
       " {'ORF': -1, 'sequence': 'CCHA'},\n",
       " {'ORF': -2, 'sequence': 'AAMH'},\n",
       " {'ORF': -3, 'sequence': 'LPC'}]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_orfs(\"ATGCATGGCAGCA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "for dirname in ALLDIRNAMES:\n",
    "    if((dirname[len(dirname) - 3:] == \"fna\") and not (dirname[:len(dirname) - 3] + \"faa\") in ALLDIRNAMES):\n",
    "        print(\"Opening up \" + dirname + \" and converting into peptide sequences...\")\n",
    "        DNAseqs = []\n",
    "        seqDescriptions = []\n",
    "        for fastaobj in readFASTA(dirname):\n",
    "            DNAseqs.append(fastaobj[\"sequence\"])\n",
    "            seqDescriptions.append(fastaobj[\"description\"])\n",
    "            \n",
    "        entries = []\n",
    "        for i in range(0, len(DNAseqs)):\n",
    "            print(\"converting \" + str(len(DNAseqs[i])) + \" base pairs from \" + seqDescriptions[i])\n",
    "            aalist = get_orfs(DNAseqs[i])\n",
    "            print(\"created \" + str(len(aalist)) + \" peptide sequences from \" + seqDescriptions[i])\n",
    "            for e in range(0, len(aalist)):\n",
    "                entries.append({\n",
    "                    \"sequence\": aalist[e][\"sequence\"],\n",
    "                    \"description\": str(seqDescriptions[i] + \" - ORF \" + str(aalist[e][\"ORF\"])) \n",
    "                })\n",
    "        \n",
    "        print(\"writing read peptides into '\" + dirname[len('genomes/'):len(dirname) - 3] + \"faa'\")\n",
    "        with open(dirname[:len(dirname) - 3] + \"faa\", 'w') as outfile:\n",
    "            for ent in entries:\n",
    "                outfile.write(\"> \" + ent[\"description\"] + \"\\n\")\n",
    "                outfile.write(ent[\"sequence\"] + \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we append all of the files ending in a \".faa\" translation to an array of files called DIRNAMES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831789.1.faa', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831790.1.faa', 'genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831788.1.faa', 'genomes/Streptomyces_lividans_TK24_GCA_000739105_1\\uf028/NZ_CP009124.1.faa', 'genomes/Streptomyces_albus_GCA_000827005_1\\uf028/NZ_CP010519.1.faa', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020816.1.faa', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020815.1.faa', 'genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1\\uf028/NC_020817.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014818.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014819.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014817.1.faa', 'genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2\\uf028/NC_014816.1.faa', 'genomes/Frankia_CcI3_GCA_000013345_1\\uf028/NC_007777.1.faa', 'genomes/Streptococcus_suis_ST3_GCA_000204625_1\\uf028/NC_015433.1.faa', 'genomes/Streptomyces_albus_GCA_001577385_1\\uf028/NZ_CP014485.1.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000071.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000022.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000042.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000060.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000079.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000062.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000051.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000034.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000067.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000041.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000072.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000039.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000083.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000094.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000014.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000032.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000003.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000026.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000080.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000023.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000065.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000007.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000013.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000048.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000081.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000054.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000100.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000047.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000078.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000005.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000031.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000040.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000028.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000077.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000044.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000084.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000085.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000025.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000049.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000058.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000033.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000008.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000068.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000055.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000019.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000016.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000099.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000074.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000037.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000012.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000096.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000052.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000053.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000063.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000024.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000090.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000092.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000093.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000059.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000004.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000061.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000020.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000086.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000010.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000064.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000015.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000070.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000097.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000075.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000091.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000095.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000088.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000066.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000098.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000011.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000002.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000006.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000056.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000043.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000036.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000050.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000030.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000069.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000076.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000087.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000035.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000029.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000018.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000089.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000057.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000021.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000038.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000009.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000001.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000017.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000073.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000082.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000045.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000046.faa', 'genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000027.faa', 'genomes/Streptomyces_albus_J1074_GCA_000359525_1\\uf028/NC_020990.1.faa']\n"
     ]
    }
   ],
   "source": [
    "DIRNAMES = []\n",
    "for dirname in os.listdir(\"genomes\"):\n",
    "    if (dirname.find(\".\") != -1):\n",
    "        if(dirname[len(dirname) - 3:] == \"faa\"):\n",
    "            DIRNAMES.append(\"genomes/\" + dirname)\n",
    "    else:\n",
    "        for filename in os.listdir(\"genomes/\" + dirname):\n",
    "            if(filename[len(filename) - 3:] == \"faa\"):\n",
    "                DIRNAMES.append(\"genomes/\" + dirname + \"/\" + filename)\n",
    "    \n",
    "        \n",
    "print(DIRNAMES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MAST method\n",
    "\n",
    "Define a method that lets us take a sequence and identify B and C proteins from it. Requires the sequence and the directory of the pre-generated MEME motif files as input, and returns a tuple containing the B matches and the C matches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mastSearch(sequence, memeDirB, memeDirC):\n",
    "    Bproteins = []\n",
    "    Cproteins = []\n",
    "    with open(\"tempseq.txt\", \"w\") as file:\n",
    "        file.write(\"> \" + \"temporary\" + \"\\n\")\n",
    "        file.write(sequence)\n",
    "        file.close()\n",
    "\n",
    "    os.system('mast -hit_list ' + memeDirB + ' tempseq.txt > tempoutB.txt')\n",
    "    os.system('mast -hit_list ' + memeDirC + ' tempseq.txt > tempoutC.txt')\n",
    "\n",
    "    with open(\"tempoutB.txt\", \"r\") as file:\n",
    "        inlines = file.readlines()\n",
    "        inlines = inlines[2:len(inlines) - 1]\n",
    "\n",
    "        \n",
    "        for line in inlines:\n",
    "            # remove ending newline character\n",
    "            line = line[:len(line) - 1]\n",
    "            params = line.split(' ')\n",
    "            while('' in params) : \n",
    "                params.remove('') \n",
    "            try:\n",
    "                newB = {\n",
    "                    \"strand\" : int(params[1]),\n",
    "                    \"motif\" : params[2],\n",
    "                    \"start\" : int(params[4]),\n",
    "                    \"end\" : int(params[5]),\n",
    "                    \"score\" : float(params[6]),\n",
    "                    \"p-value\" : float(params[7])\n",
    "                }\n",
    "                Bproteins.append(newB)\n",
    "            except:\n",
    "                print(\"error in parsing line - \" + line)\n",
    "                print(\"params: \" + str(params))\n",
    "        file.close()\n",
    "    with open(\"tempoutC.txt\", \"r\") as file:\n",
    "        inlines = file.readlines()\n",
    "        inlines = inlines[2:len(inlines) - 1]\n",
    "\n",
    "        \n",
    "        for line in inlines:\n",
    "            # remove ending newline character\n",
    "            line = line[:len(line) - 1]\n",
    "            params = line.split()\n",
    "            while('' in params) : \n",
    "                params.remove('') \n",
    "            try:\n",
    "                newC = {\n",
    "                    \"strand\" : int(params[1]),\n",
    "                    \"motif\" : params[2],\n",
    "                    \"start\" : int(params[4]),\n",
    "                    \"end\" : int(params[5]),\n",
    "                    \"score\" : float(params[6]),\n",
    "                    \"p-value\" : float(params[7])\n",
    "                }\n",
    "                Cproteins.append(newC)\n",
    "            except:\n",
    "                print(\"error in parsing line - \" + line)\n",
    "                print(\"params: \" + str(params))\n",
    "        file.close()\n",
    "\n",
    "    os.remove(\"tempseq.txt\")\n",
    "    os.remove(\"tempoutB.txt\")\n",
    "    os.remove(\"tempoutC.txt\")\n",
    "\n",
    "    return (Bproteins, Cproteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pattern Matching\n",
    "Takes in 6 sequences corresponding to 6 ORFs of a given sequence, and then identifies B/C gene clusters within them. Then, use the python regular expression library to identify precursor peptides matching the regular expression pattern at the top of the script. The function returns a list of matched proteins, which have a specific sequence, ORF, nearest B/C cluster, and range within the overall sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Some helper functions\n",
    "def isOverlapping(start1, end1, start2, end2):\n",
    "    if (start1 <= start2) and (end1 >= start2):\n",
    "        return True\n",
    "    if(start2 <= start1) and (end2 >= start1):\n",
    "        return True\n",
    "    \n",
    "    \n",
    "    return False\n",
    "\n",
    "def adjustRangeByORF(ORF, length, start, end):\n",
    "    if ORF == 2:\n",
    "        start += 1\n",
    "        end += 1\n",
    "    elif ORF == 3:\n",
    "        start += 1\n",
    "        end += 1\n",
    "    elif ORF == -1:\n",
    "        temp = start\n",
    "        start = length - end\n",
    "        end = length - temp\n",
    "    elif ORF == -2:\n",
    "        temp = start\n",
    "        start = length - end - 1\n",
    "        end = length - temp - 1\n",
    "    elif ORF == -3:\n",
    "        temp = start\n",
    "        start = length - end - 2\n",
    "        end = length - temp - 2\n",
    "    \n",
    "    return [start, end]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def patternMatch(sequenceORFs, pattern, filename):\n",
    "    Aproteins = []\n",
    "    Bproteins = []\n",
    "    Cproteins = []\n",
    "    \n",
    "    \n",
    "    ## generate all of the Bs and Cs, and label them by ORF\n",
    "    ORFs = [1, 2, 3, -1, -2, -3]\n",
    "    ORF = 0\n",
    "    for pair in sequenceORFs:\n",
    "        overallSequence = pair[\"sequence\"]\n",
    "        pair = mastSearch(overallSequence, \"motifs/memeb.txt\", \"motifs/memec.txt\")\n",
    "        \n",
    "        newBs = pair[0]\n",
    "        newCs = pair[1]\n",
    "        \n",
    "        for b in newBs:\n",
    "            b[\"ORF\"] = ORFs[ORF]\n",
    "            prange = adjustRangeByORF(ORFs[ORF], len(overallSequence), b[\"start\"], b[\"end\"])\n",
    "            b[\"start\"] = prange[0]\n",
    "            b[\"end\"] = prange[1]\n",
    "            \n",
    "            \n",
    "        for c in newCs:\n",
    "            c[\"ORF\"] = ORFs[ORF]\n",
    "            prange = adjustRangeByORF(ORFs[ORF], len(overallSequence), c[\"start\"], c[\"end\"])\n",
    "            c[\"start\"] = prange[0]\n",
    "            c[\"end\"] = prange[1]\n",
    "        \n",
    "        Bproteins.extend(newBs)\n",
    "        Cproteins.extend(newCs)\n",
    "        \n",
    "        ORF += 1\n",
    "        \n",
    "    ## create all of the cluster points, and give them a score\n",
    "    ORF = 0\n",
    "    for pair in sequenceORFs:\n",
    "        overallSequence = pair[\"sequence\"]\n",
    "        description = filename[8:len(filename) - 4] + \" - \" + pair[\"description\"]\n",
    "        # find all matches in protein that match\n",
    "        matchIter = re.finditer(pattern, overallSequence)\n",
    "        done_looping = False\n",
    "        while not done_looping:\n",
    "            try:\n",
    "                match = next(matchIter)\n",
    "            except StopIteration:\n",
    "                done_looping = True\n",
    "            else:\n",
    "\n",
    "                # get the correct range based on span\n",
    "                indices = list(match.span())\n",
    "                indices = adjustRangeByORF(ORFs[ORF], len(overallSequence), indices[0], indices[1])\n",
    "\n",
    "                # make the ranking calculation and find closest B and C\n",
    "                rank = 0\n",
    "                start = indices[0]\n",
    "                end = indices[1]\n",
    "\n",
    "                term1 = 0\n",
    "                closest = float(\"inf\")\n",
    "                closestB = None\n",
    "                if len(Cproteins) != 0:\n",
    "                    for prot in Bproteins:\n",
    "                        if isOverlapping(start, end, prot[\"start\"], prot[\"end\"]):\n",
    "                            # print(str(start) + \"-\" + str(end) + \" |B at \" + str(prot[\"start\"]) + \"-\" + str(prot[\"end\"]))\n",
    "                            # print(prot)\n",
    "                            continue\n",
    "                        diffsquared = (prot[\"start\"] - start) ** 2\n",
    "                        if diffsquared < closest:\n",
    "                            closestB = prot\n",
    "                            closest = diffsquared\n",
    "                        term1 += (1.0 / float(diffsquared))\n",
    "                            \n",
    "\n",
    "                term2 = 0\n",
    "                closest = float(\"inf\")\n",
    "                closestC = None\n",
    "                if len(Bproteins) != 0:\n",
    "\n",
    "                    for prot in Cproteins:\n",
    "                        if isOverlapping(start, end, prot[\"start\"], prot[\"end\"]):\n",
    "                            # print(str(start) + \"-\" + str(end) + \" |C at \" + str(prot[\"start\"]) + \"-\" + str(prot[\"end\"]))\n",
    "                            # print(prot)\n",
    "                            continue\n",
    "                        diffsquared = (prot[\"start\"] - start) ** 2\n",
    "\n",
    "                        if diffsquared < closest:\n",
    "                            closestC = prot\n",
    "                            closest = diffsquared\n",
    "                        term2 += (1.0 / float(diffsquared))\n",
    "\n",
    "                rank = term1 * term2\n",
    "                if rank == 0:\n",
    "                    continue\n",
    "                else:\n",
    "                    rank = -1 * math.log(rank, 10)\n",
    "\n",
    "                # append the protein to the list of proteins\n",
    "                Aproteins.append({\n",
    "                    \"description\": description,\n",
    "                    \"sequence\": match.group(0),\n",
    "                    \"searchPattern\": match.re.pattern,\n",
    "                    \"searchRange\": indices,\n",
    "                    \"overallLength\": len(overallSequence),\n",
    "                    \"rank\": rank,\n",
    "                    \"closestB\": closestB,\n",
    "                    \"closestC\": closestC,\n",
    "                    \"ORF\": ORFs[ORF],\n",
    "                    \"genome\": description[:description.index(\"/\")]\n",
    "                    ## \"overallString\": match.string\n",
    "                })\n",
    "                \n",
    "        ORF += 1\n",
    "    return Aproteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some unit testing..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 6 objects from FASTA file genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1/NZ_LN831788.1.faa\n",
      "The A proteins...\n",
      "[3417, 3486] - 8.350842598439872 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 3291, 'end': 3311, 'score': 126.08, 'p-value': 3.76e-05, 'ORF': -3} - \n",
      "[3927, 3987] - 9.25293317339028 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3809, 'end': 3823, 'score': 544.4, 'p-value': 9.49e-05, 'ORF': 2} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 3291, 'end': 3311, 'score': 126.08, 'p-value': 3.76e-05, 'ORF': -3} - \n",
      "[4833, 4893] - 10.804559997750356 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 4364, 'end': 4378, 'score': 584.88, 'p-value': 7.99e-05, 'ORF': 2} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 3291, 'end': 3311, 'score': 126.08, 'p-value': 3.76e-05, 'ORF': -3} - \n",
      "[9620, 9686] - 10.909920269872357 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 10323, 'end': 10330, 'score': 1052.48, 'p-value': 3.14e-05, 'ORF': 1} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 8497, 'end': 8521, 'score': 974.08, 'p-value': 6.4e-08, 'ORF': 2} - \n",
      "[13868, 13941] - 10.057670865736075 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 13516, 'end': 13523, 'score': 902.46, 'p-value': 7e-05, 'ORF': -1} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 13198, 'end': 13218, 'score': 50.64, 'p-value': 4.92e-05, 'ORF': -1} - \n",
      "[14561, 14614] - 9.938943546558415 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 13516, 'end': 13523, 'score': 902.46, 'p-value': 7e-05, 'ORF': -1} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 14717, 'end': 14740, 'score': 358.08, 'p-value': 8.2e-06, 'ORF': -3} - \n",
      "[19753, 19787] - 8.848916837029131 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 20156, 'end': 20163, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -2} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 19683, 'end': 19703, 'score': 757.89, 'p-value': 3.44e-06, 'ORF': -2} - \n",
      "[22083, 22126] - 11.445862159931622 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 23224, 'end': 23238, 'score': 1020.04, 'p-value': 1.13e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[9722, 9801] - 10.870791594973769 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 10323, 'end': 10330, 'score': 1052.48, 'p-value': 3.14e-05, 'ORF': 1} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 8497, 'end': 8521, 'score': 974.08, 'p-value': 6.4e-08, 'ORF': 2} - \n",
      "[13920, 13966] - 10.176884387306645 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 13516, 'end': 13523, 'score': 902.46, 'p-value': 7e-05, 'ORF': -1} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 13198, 'end': 13218, 'score': 50.64, 'p-value': 4.92e-05, 'ORF': -1} - \n",
      "[16918, 16977] - 10.856557892143236 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 17503, 'end': 17517, 'score': 534.28, 'p-value': 9.9e-05, 'ORF': 3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 15776, 'end': 15796, 'score': 31.78, 'p-value': 5.26e-05, 'ORF': 3} - \n",
      "[19504, 19580] - 9.932907506519834 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 20156, 'end': 20163, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -2} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 19683, 'end': 19703, 'score': 757.89, 'p-value': 3.44e-06, 'ORF': -2} - \n",
      "[21396, 21480] - 10.730895594073461 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 20156, 'end': 20163, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -2} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[22030, 22085] - 11.43751588753232 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 23224, 'end': 23238, 'score': 1020.04, 'p-value': 1.13e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[27199, 27278] - 9.999576530786733 - {'strand': 1, 'motif': 'QPFAAHCWVZAGGVVLBDSPE', 'start': 26993, 'end': 27013, 'score': -557.25, 'p-value': 7.84e-05, 'ORF': 3} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 27948, 'end': 27972, 'score': -982.88, 'p-value': 7.64e-05, 'ORF': -1} - \n",
      "[3264, 3306] - 9.8944662676333 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 2924, 'end': 2947, 'score': -302.4, 'p-value': 8.64e-05, 'ORF': 1} - \n",
      "[10667, 10702] - 10.154218717921307 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 11009, 'end': 11023, 'score': 696.2, 'p-value': 4.95e-05, 'ORF': -2} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 11303, 'end': 11327, 'score': -1019.12, 'p-value': 8.56e-05, 'ORF': -1} - \n",
      "[11719, 11783] - 10.56984833029787 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 11009, 'end': 11023, 'score': 696.2, 'p-value': 4.95e-05, 'ORF': -2} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 11303, 'end': 11327, 'score': -1019.12, 'p-value': 8.56e-05, 'ORF': -1} - \n",
      "[16422, 16473] - 9.481696449773636 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 16310, 'end': 16317, 'score': 902.46, 'p-value': 7e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 15776, 'end': 15796, 'score': 31.78, 'p-value': 5.26e-05, 'ORF': 3} - \n",
      "[17325, 17376] - 10.147498717151883 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 17503, 'end': 17517, 'score': 534.28, 'p-value': 9.9e-05, 'ORF': 3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 18684, 'end': 18707, 'score': -157.92, 'p-value': 5.28e-05, 'ORF': 1} - \n",
      "[20496, 20564] - 10.29602029088615 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 20156, 'end': 20163, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -2} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[22850, 22886] - 10.928025555398639 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 23224, 'end': 23238, 'score': 1020.04, 'p-value': 1.13e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[23640, 23685] - 10.553510442933394 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 23397, 'end': 23404, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 25264, 'end': 25284, 'score': 541.0, 'p-value': 8.06e-06, 'ORF': 3} - \n",
      "[26307, 26382] - 10.666094455132122 - {'strand': 1, 'motif': 'QPFAAHCWVZAGGVVLBDSPE', 'start': 26993, 'end': 27013, 'score': -557.25, 'p-value': 7.84e-05, 'ORF': 3} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 25482, 'end': 25506, 'score': -656.72, 'p-value': 2.67e-05, 'ORF': -2} - \n",
      "[25959, 25997] - 9.994617772645759 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 25583, 'end': 25590, 'score': 1098.64, 'p-value': 2.43e-05, 'ORF': 1} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 25482, 'end': 25506, 'score': -656.72, 'p-value': 2.67e-05, 'ORF': -2} - \n",
      "[22980, 23052] - 10.618764111868702 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 23224, 'end': 23238, 'score': 1020.04, 'p-value': 1.13e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[20771, 20846] - 10.418296612714819 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 20156, 'end': 20163, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -2} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 21092, 'end': 21115, 'score': -240.48, 'p-value': 7.01e-05, 'ORF': 2} - \n",
      "[16896, 16963] - 10.851330627955184 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 16310, 'end': 16317, 'score': 902.46, 'p-value': 7e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 15776, 'end': 15796, 'score': 31.78, 'p-value': 5.26e-05, 'ORF': 3} - \n",
      "[6619, 6666] - 10.658573386144834 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 7380, 'end': 7394, 'score': 1414.72, 'p-value': 1.57e-06, 'ORF': -2} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 7136, 'end': 7160, 'score': -874.16, 'p-value': 5.41e-05, 'ORF': -3} - \n",
      "[2264, 2323] - 11.157821747947574 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 2924, 'end': 2947, 'score': -302.4, 'p-value': 8.64e-05, 'ORF': 1} - \n",
      "[452, 516] - 11.147947532922126 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 684, 'end': 704, 'score': 267.53, 'p-value': 2.25e-05, 'ORF': 2} - \n",
      "[24874, 24935] - 10.255690091653157 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 25528, 'end': 25542, 'score': 1434.96, 'p-value': 1.41e-06, 'ORF': 2} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 25264, 'end': 25284, 'score': 541.0, 'p-value': 8.06e-06, 'ORF': 3} - \n",
      "[20143, 20198] - 11.326397866820223 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 17503, 'end': 17517, 'score': 534.28, 'p-value': 9.9e-05, 'ORF': 3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 19683, 'end': 19703, 'score': 757.89, 'p-value': 3.44e-06, 'ORF': -2} - \n",
      "[19781, 19841] - 9.077600880977753 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 20156, 'end': 20163, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -2} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 19683, 'end': 19703, 'score': 757.89, 'p-value': 3.44e-06, 'ORF': -2} - \n",
      "[14387, 14440] - 10.373572833602477 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 13516, 'end': 13523, 'score': 902.46, 'p-value': 7e-05, 'ORF': -1} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 14717, 'end': 14740, 'score': 358.08, 'p-value': 8.2e-06, 'ORF': -3} - \n",
      "[11128, 11198] - 8.604334616689172 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 11009, 'end': 11023, 'score': 696.2, 'p-value': 4.95e-05, 'ORF': -2} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 11303, 'end': 11327, 'score': -1019.12, 'p-value': 8.56e-05, 'ORF': -1} - \n",
      "[6850, 6918] - 8.99771438329939 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 6605, 'end': 6619, 'score': 574.76, 'p-value': 8.35e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 6660, 'end': 6683, 'score': 162.0, 'p-value': 1.69e-05, 'ORF': 1} - \n",
      "[5307, 5374] - 11.13713995032291 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 4364, 'end': 4378, 'score': 584.88, 'p-value': 7.99e-05, 'ORF': 2} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 6660, 'end': 6683, 'score': 162.0, 'p-value': 1.69e-05, 'ORF': 1} - \n",
      "[2464, 2520] - 10.815948746183112 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 2924, 'end': 2947, 'score': -302.4, 'p-value': 8.64e-05, 'ORF': 1} - \n",
      "[23802, 23859] - 10.87826128671755 - {'strand': 4, 'motif': 'CLPRSIAL', 'start': 23397, 'end': 23404, 'score': 960.16, 'p-value': 5.18e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 25264, 'end': 25284, 'score': 541.0, 'p-value': 8.06e-06, 'ORF': 3} - \n",
      "[11479, 11533] - 9.678552230657056 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 11009, 'end': 11023, 'score': 696.2, 'p-value': 4.95e-05, 'ORF': -2} - {'strand': 2, 'motif': 'LLAQPVVEACLAIPTWDLAEGGQDR', 'start': 11303, 'end': 11327, 'score': -1019.12, 'p-value': 8.56e-05, 'ORF': -1} - \n",
      "[7332, 7394] - 9.245850626737663 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 6605, 'end': 6619, 'score': 574.76, 'p-value': 8.35e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 7440, 'end': 7460, 'score': -5.94, 'p-value': 6e-05, 'ORF': -3} - \n",
      "[6227, 6269] - 10.099095668005061 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 6605, 'end': 6619, 'score': 574.76, 'p-value': 8.35e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 6660, 'end': 6683, 'score': 162.0, 'p-value': 1.69e-05, 'ORF': 1} - \n",
      "[1889, 1960] - 11.555148018729477 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 1, 'motif': 'DALFTGQGGDNVFCQVPSAAPALD', 'start': 2924, 'end': 2947, 'score': -302.4, 'p-value': 8.64e-05, 'ORF': 1} - \n",
      "[1784, 1859] - 11.622525400306428 - {'strand': 2, 'motif': 'LRARGVYAHLVIGVR', 'start': 3592, 'end': 3606, 'score': 665.84, 'p-value': 5.65e-05, 'ORF': -3} - {'strand': 3, 'motif': 'ARDSSAILVELSGGLDSSIVA', 'start': 684, 'end': 704, 'score': 267.53, 'p-value': 2.25e-05, 'ORF': 2} - \n"
     ]
    }
   ],
   "source": [
    "pair = readFASTA(DIRNAMES[2])\n",
    "A = patternMatch(pair, PATTERN, \"file/thinggenome/file\")\n",
    "    \n",
    "print(\"The A proteins...\")\n",
    "for prot in A:\n",
    "    print(str(prot[\"searchRange\"]) + \" - \" + str(prot[\"rank\"]) + \" - \" + str(prot[\"closestB\"]) + \" - \" + str(prot[\"closestC\"]) + \" - \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The Search\n",
    "Run the pattern matching method on all of the directoried genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 6 objects from FASTA file genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1/NZ_LN831789.1.faa\n",
      "Read 6 objects from FASTA file genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1/NZ_LN831790.1.faa\n",
      "Read 6 objects from FASTA file genomes/Streptomyces_leeuwenhoekii_GCA_001013905_1/NZ_LN831788.1.faa\n",
      "Read 6 objects from FASTA file genomes/Streptomyces_lividans_TK24_GCA_000739105_1/NZ_CP009124.1.faa\n",
      "Read 6 objects from FASTA file genomes/Streptomyces_albus_GCA_000827005_1/NZ_CP010519.1.faa\n",
      "Read 6 objects from FASTA file genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1/NC_020816.1.faa\n",
      "Read 6 objects from FASTA file genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1/NC_020815.1.faa\n",
      "Read 6 objects from FASTA file genomes/Xanthomonas_citri_citri_Aw12879_GCA_000349225_1/NC_020817.1.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2/NC_014818.1.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2/NC_014819.1.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2/NC_014817.1.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_excentricus_CB_48_GCA_000175215_2/NC_014816.1.faa\n",
      "Read 6 objects from FASTA file genomes/Frankia_CcI3_GCA_000013345_1/NC_007777.1.faa\n",
      "Read 6 objects from FASTA file genomes/Streptococcus_suis_ST3_GCA_000204625_1/NC_015433.1.faa\n",
      "Read 6 objects from FASTA file genomes/Streptomyces_albus_GCA_001577385_1/NZ_CP014485.1.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000071.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000022.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000042.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000060.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000079.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000062.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000051.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000034.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000067.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000041.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000072.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000039.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000083.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000094.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000014.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000032.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000003.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000026.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000080.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000023.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000065.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000007.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000013.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000048.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000081.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000054.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000100.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000047.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000078.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000005.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000031.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000040.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000028.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000077.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000044.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000084.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000085.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000025.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000049.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000058.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000033.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000008.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000068.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000055.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000019.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000016.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000099.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000074.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000037.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000012.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000096.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000052.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000053.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000063.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000024.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000090.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000092.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000093.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000059.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000004.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000061.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000020.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000086.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000010.faa\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000064.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000015.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000070.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000097.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000075.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000091.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000095.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000088.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000066.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000098.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000011.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000002.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000006.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000056.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000043.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000036.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000050.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000030.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000069.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000076.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000087.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000035.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000029.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000018.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000089.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000057.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000021.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000038.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000009.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000001.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000017.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000073.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000082.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000045.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000046.faa\n",
      "Read 6 objects from FASTA file genomes/Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000027.faa\n",
      "Read 6 objects from FASTA file genomes/Streptomyces_albus_J1074_GCA_000359525_1/NC_020990.1.faa\n",
      "Found 24740 that satisfy the pattern: M[A-Z]{15,45}T[A-Z][A-Z]{6,8}[DE][A-Z]{5,30}\\*\n"
     ]
    }
   ],
   "source": [
    "matchedProteins = []\n",
    "for filename in DIRNAMES:\n",
    "    readSequences = readFASTA(filename)\n",
    "    if(not len(readSequences) == 6):\n",
    "        print(\"Error: sequence in file \" + filename + \" does not have 6 sequences for 6 ORFs\")\n",
    "        print(\"Instead, it has \" + str(len(readSequences)))\n",
    "        raise RuntimeError\n",
    "    matchedProteins.extend(patternMatch(readSequences, PATTERN, filename))\n",
    "    \n",
    "\n",
    "print(\"Found \" + str(len(matchedProteins)) + \" that satisfy the pattern: \" + PATTERN)\n",
    "# for match in matchedProteins:\n",
    "#     print(match[\"sequence\"] + \", found in range \" + str(match[\"searchRange\"]))\n",
    "#     print(\"description: \" + match[\"description\"])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'description': 'Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028/NZ_LN831790.1 -  gi|824057350|ref|NZ_LN831790.1| Streptomyces leeuwenhoekii genome assembly sleC34, chromosome : chromosome - ORF 1',\n",
       " 'sequence': 'MSPCTKSRRGWPSSRLSGFPAYVSLSSTVIRGAPGASEASSART*',\n",
       " 'searchPattern': 'M[A-Z]{15,45}T[A-Z][A-Z]{6,8}[DE][A-Z]{5,30}\\\\*',\n",
       " 'searchRange': [190961, 191006],\n",
       " 'overallLength': 2634631,\n",
       " 'rank': 11.562712028666043,\n",
       " 'closestB': {'strand': 4,\n",
       "  'motif': 'CLPRSIAL',\n",
       "  'start': 189626,\n",
       "  'end': 189633,\n",
       "  'score': 1029.4,\n",
       "  'p-value': 3.57e-05,\n",
       "  'ORF': 2},\n",
       " 'closestC': {'strand': 3,\n",
       "  'motif': 'ARDSSAILVELSGGLDSSIVA',\n",
       "  'start': 191811,\n",
       "  'end': 191831,\n",
       "  'score': 427.84,\n",
       "  'p-value': 1.24e-05,\n",
       "  'ORF': -3},\n",
       " 'ORF': 1,\n",
       " 'genome': 'Streptomyces_leeuwenhoekii_GCA_001013905_1\\uf028'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matchedProteins[100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('matches.json', 'w') as outfile:\n",
    "    json.dump(matchedProteins, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "lassopeptides = []\n",
    "with open('matches.json', 'r') as storedfile:\n",
    "    lassopeptides = json.loads(storedfile.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing output to 'matches.csv'\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "print(\"Writing output to 'matches.csv'\")\n",
    "pd.read_json(\"matches.json\").to_csv(\"matches.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24740\n"
     ]
    }
   ],
   "source": [
    "validpeptides = []\n",
    "for peptide in lassopeptides:\n",
    "    if(peptide[\"closestB\"] != None):\n",
    "        validpeptides.append(peptide)\n",
    "    \n",
    "\n",
    "print(len(validpeptides))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'description': 'Streptomyces_lividans_TK24_GCA_000739105_1\\uf028/NZ_CP009124.1 -  gi|749298793|ref|NZ_CP009124.1| Streptomyces lividans TK24, complete genome - ORF -1',\n",
       " 'sequence': 'MGVPPRERVRAWGSVQRPAVGLVPPTTAYSGVRLPEVSLCHG*',\n",
       " 'searchPattern': 'M[A-Z]{15,45}T[A-Z][A-Z]{6,8}[DE][A-Z]{5,30}\\\\*',\n",
       " 'searchRange': [1184125, 1184168],\n",
       " 'overallLength': 2781761,\n",
       " 'rank': 10.439802636937632,\n",
       " 'closestB': {'strand': 2,\n",
       "  'motif': 'LRARGVYAHLVIGVR',\n",
       "  'start': 1184737,\n",
       "  'end': 1184751,\n",
       "  'score': 1111.12,\n",
       "  'p-value': 7.29e-06,\n",
       "  'ORF': 2},\n",
       " 'closestC': {'strand': 3,\n",
       "  'motif': 'ARDSSAILVELSGGLDSSIVA',\n",
       "  'start': 1184425,\n",
       "  'end': 1184445,\n",
       "  'score': -147.39,\n",
       "  'p-value': 9.79e-05,\n",
       "  'ORF': 3},\n",
       " 'ORF': -1,\n",
       " 'genome': 'Streptomyces_lividans_TK24_GCA_000739105_1\\uf028'}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validpeptides[4636]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sort by rank..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getRank(precursor):\n",
    "    return precursor[\"rank\"]\n",
    "\n",
    "validpeptides.sort(key=getRank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAdMklEQVR4nO3deXxU533v8c8jgcQmtG8IC0kIkEGYAmIzGPBuHLu2c53WbhYvcXHSNG3SXt/i5g/7NmnaG9/r26Rp3dhZ7Zs4xjap7SSOwQ6LjdnEvkoIgUAS2le0a+a5f8zgqBhZIM3MmTP6vl8vXhqdGXF+z5zR93X0O885x1hrERER94tyugAREQkMBbqISIRQoIuIRAgFuohIhFCgi4hEiDGhXFlKSorNyckJ5SpFRFxv7969Ddba1KFeF9JAz8nJobi4OJSrFBFxPWNMxZW8Ti0XEZEIoUAXEYkQCnQRkQihQBcRiRAKdBGRCKFAFxGJEAp0EZEIoUAXEQmi6pYunt1YwumGjqCvS4EuIhJE51u7+d7vy6hoVKCLiLic7yZCxpigr0mBLiISRBdvChf8OFegi4gE1cWbfIZgB12BLiISTH/YQ1fLRUTE1ay92EMP/roU6CIiQdTr8QIwNjr4catAFxEJoubOPgASJ4wN+roU6CIiQdTS2QtAwoSYoK9LgS4iEkTNHb499Pjx2kMXEXG1lq5eJsWOIWaMeugiIq7W0tlHQgj656BAFxEJqubOXhJD0D8HBbqISFA1aw9dRCQyNHdoD11ExPWstdS1d5MWFxuS9SnQRUSC5HRDB919XqYlTwjJ+hToIiJB8vaRGgBumZ0ekvUp0EVEgmRbaT0FGXFkxo8PyfoU6CIiQdDW3UdxRTM3FqSFbJ0KdBGRIHi1uBKP13JriNotoEAXEQm4c02dPLuxhOX5ySzITgzZehXoIiIB9u3fHscYw3funxfS9SrQRUQC6EhVK28fqeGR5TlkJYTmYOhFCnQRkQDp7vPw5IbDxI8fy2M35IV8/WNCvkYRkQhkreWJ1w5xpLqVFz5fFJLrn19Ke+giIgHw0s4K3jpYzd/eOjNkJxJdSoEuIjJC7xyt4X++dYybC9L48up8x+pQoIuIjMDmkjq++vJ+CqdM5l8e+COio4xjtSjQRUSGaf2eczz2s2LyUyfxo4cXETcu9H3zgXRQVETkKllr+b/vnuR7753khhkp/PtnFzge5qBAFxG5Kn0eL3/3+iE27Kvi/oVT+cf7CokdE+10WcAVtFyMMT82xtQZY44MWPaMMeaEMeaQMeZXxpiE4JYpIuK8c02dPP7SXjbsq+Lrt8zkmfuvC5swhyvrof8UuOOSZZuAQmvtdUAp8GSA6xIRCRser+XFHWe45dmtvH+ynm/dW8hf3zIDY5w7AHo5Q7ZcrLXbjDE5lyzbOODbncD9gS1LRMR51lrePV7HM++coLT2AqtmpvJPn57LlBCf0n+lAtFDfxR4ZbAnjTFrgbUA2dnZAVidiEjw7TjVyLObSthzppnclIn864Pz+dTcTKIcnJY4lBEFujHmG0A/8PPBXmOtfR54HqCoqMiOZH0iIsHW3t3H028e4/V9laRPjuWb9xbywKJrGBsd/rO8hx3oxpiHgbuAm621CmoRcbXefi/ri8/xvfdO0tjRy1dunM5Xb5rBuLHhc9BzKMMKdGPMHcD/AFZZazsDW5KISOj0eby8vreS728uo7K5i6Jpifzg8wuZH8IbUwTKkIFujHkZWA2kGGMqgafwzWqJBTb5j/LutNZ+KYh1iogE1IWefjbsq+SH75/mbFMn86bG8817C1k9MzXsZq9cqSuZ5fLgZRb/KAi1iIgEXWVzJy/uqODl3Wdp7+5n3jUJPHX3bG4qSHNtkF+kM0VFJOJZa9lR3shLOyrYeKwWay1rCjN5dEUOC7ITXR/kFynQRSRiNXf08ubBav7fzgpO1l0gYcJYHluRy0PX54TtXPKRUKCLSETp6vWw8VgNbx2sZmtpPX0eS2HWZJ65/zrunjfFVbNWrpYCXURcz+O17CxvZMO+Kn535DwdvR4y48fx0LIc/tvCqVybOdnpEkNCgS4irlVS086G/ZW8sb+amrZu4mLH8KnrMrlv/lSW5CaF9VmdwaBAFxFXqWvr5o0D1WzYX8Xx821ERxlWzUzlG5+6lltnp0d0S2UoCnQRCXudvf28c7SGDfuq2F7WgNfCvKnxPH33bO6aN4WUSbFOlxgWFOgiEpb6PF42n6jj3eO1vH2khvbufrISxvMXq/O5d34W+WmTnC4x7CjQRSSslNW186v9VbxaXEldew9x48Zwc0EaDyzOZnHO6OuLXw0Fuog4rqqli7cPn+c/D1RxpKqNKAOrZqby7SXTWDUr1RVXOgwHCnQRccTJ2nY2l9Sx6Vgte840A76++FN3z+au66aQGqe++NVSoItISPR5vBSfaWZLaR3vHa+jrO4CADPSJvHE7bNYU5hBXqr64iOhQBeRoDnf2sXWkno2l9SxvayRCz39jI02LMpJ4qFl07h1dgYZ8eOcLjNiKNBFJGAG7oVvLannRE07AJnx47h73hRWz0pleX4Kk2IVPcGgd1VERqSmtZstJXVsKalne1kD7T39jIny7YU/uaaA1bPSmJk+KWKuaBjOFOgiclX6PF72VjSzpaSeLSV1/2Uv/K55mayamcby/GTixo11uNLRR4EuIkM639rF9rJGNpfUsa20nvZu3174wmmJrFtTwOpZqcxKj9NeuMMU6CLyMa1dfewsb2R7WQPbyxo4Vd8BQGpcLGsKM7ipII0VM1LVCw8z2hoiQk+/h70VzXxY1sgHZQ0cqmzBa2H82GiW5CXx4OJsrp+eQkFGnM7UDGMKdJFRyOu1HDvfxvayBj4oa2DPmSa6+7xERxnmTY3nL2/MZ3l+CvOzE4kZo7M03UKBLjJKVLd08cHJBnaWN7K1tJ7Gjl7Ad2LPA4uyWZGfwpK8JB3MdDEFukiE6u33svt0E9tO1rOt9A9zwpMmxnDDjBRWzkhlxYwU0ifrxJ5IoUAXiRDWWk7WXWBbaT0fnmpkV3kjHb0eYqKjWDAtgSfXFHBjQRoz0jQnPFIp0EVc7FxTJzs+mo3SSMOFHgByUyZy7/wsVs/yzQmfEKNf9dFAW1nERbp6Pew508SWknreO1FLRWMn8Ic2yrK8ZFbNSiUzfrzDlYoTFOgiYazf4+VIdRvbSn2n1e8/20Kvx0tMdBQrZqTw8PU5LM9PIT91kqYTigJdJJx4vJYjVa3sOt3IjlON7D7dREevB2NgzpTJPLI8h+vzU1iUk6g2inyMPhEiDvJ6LeUNF9h9upkPT/nmhLd09gEwPdXXB1+al8zy/BSSJsY4XK2EOwW6SAhdPKHn9yfq2HOmiYPnWmjr7gcgLS6WmwvSWTkzhWXTk0mL03RCuToKdJEgstZS3tDBntNNvO+/LkpLZx/GQEHGZD513RTmZyewKCeJnOQJmk4oI6JAFwkgay1nGjvZWe7rge8sb6Su3TeV8OIe+NK8JG4qSCN5ku6ZKYGlQBcZAWst55q62FHewM7yJnacaqSmrRuAlEmxLJuezNK8JJbkJjE9VSf0SHAp0EWuUm1bN++fbODDUw3sPNVIdevFAI9hSV4yS/OSWZaXzPTUiQpwCSkFusgQWjp72V7WyI7yBj4sa6S8wXdt8KSJMSzNS+JL/gDP1yn14jAFusgluvt8Z2PuOOU7pf5wVSteCxNiolmal+y7Nnh+MtdmTNbJPBJWFOgy6rV397HvbAt7K5rZWd7IAf/ZmB9dG/ymGayelcrcrHjGRuva4BK+FOgy6lzo6WfPmSZ2nmpkb0UzB8610O+1RBkozIrn4eU5LMtLZnFuEhN1izVxEX1aJeJ193nYd9Z3e7VtJ+s5Vt1Gv9cSEx3F7CmTWbsyj+unpzDvmnjd3EFcTYEuEefiBa0u9sD3nGmip99LlIGF0xJ5fJUvwBdkJzI+JtrpckUCZshAN8b8GLgLqLPWFvqXJQGvADnAGeBPrLXNwStTZHBer+VETTsfnmpgx6lG9pxp+uh0+lnpcfzZkmyWT09hcV4Sk7UHLhHsSvbQfwp8H3hxwLJ1wHvW2n82xqzzf/93gS9P5PKqWrrYfKKOLSV17D79hwDPTZnInXMzWZ6fwtK8ZFLjdDamjB5DBrq1dpsxJueSxfcAq/2PfwZsQYEuQVTd0sX2sgaKzzSz63QjZ/w3drgmaTx3zs1kcW4Sy6Yn68YOMqoNt4eebq09739cA6QP9kJjzFpgLUB2dvYwVyejTU1rNzvKG9hV3sSO8saP7swTP34si3KS+NzSaayelarT6UUGGPFBUWutNcbYT3j+eeB5gKKiokFfJ6Nbn8fLgXMtbDpWy5aSOkprLwAQN24MS3KT+cKyHJbnJzMzLU4n84gMYriBXmuMybTWnjfGZAJ1gSxKRofmjl62ltbz7vFatpbU097Tz9how+LcJO5fOJXl+SkUZEwmWgEuckWGG+hvAg8B/+z/+kbAKpKI1dXr4cC5Fj4oq2dbaQNHqlux1ndRqzVzM1g9K40bZqRoLrjIMF3JtMWX8R0ATTHGVAJP4Qvy9caYLwIVwJ8Es0hxp4sn9Ow81cgHZQ0cqmyl32uJjjIszE7kazfPZOXMFOZNTVAbRSQArmSWy4ODPHVzgGuRCFBW186mY3VsLa1j/9mWj07omTs1gbUr8yjKSWThtCTix2svXCTQdKaojIi1liNVbbxxoIp3j9d+NJ3w2szJfHbJNFbMSKYoRyf0iISCAl2uWp/Hy54zTWw+UcfbR2qobO4iJjqK5fnJPLoil1tnp2s+uIgDFOhyRVq7+nj/ZD2bjtWy+UQdbd2+GSnL81P4q5tmcPucDOInaC9cxEkKdBlUaW07W0rq+P2JOvacacbjtSRNjOG2ORncOjud5fkpTNLlZUXChn4b5SMer+VgZQtbSurZeLSGEzXtABRkxLF2ZR43FaSxIDtR88JFwpQCfZTr7vOwvayBjUdree9ELQ0XeokysCA7kafvns1tczKYkqB+uIgbKNBHoeaOXn5/oo6Nx2rYVtpAV5+HuNgxrJqVyq2z01k5I5XEiTFOlykiV0mBPkp093l452gNr++rYntZAx6vJWPyOO5fOJVbZ6ezNC+ZmDG6X6aImynQI5i1lsNVrbx1sJrX9lbS3NlHVsJ4Hl+Zxx2FGczNiteVCkUiiAI9ApXXX2DDvireOlRNRWMnY6MNN85K4/PLprF8eopOsxeJUAr0CNHS2ctbB6v51f4q9p1tIcrA9dNT+IvV07ljTqbmiIuMAgp0F6tp7eY3h8/zztEa9lb45onPSo9j3ZoCPj0/i7TJ45wuUURCSIHuMlUtXWw6WsNvD9ewp6IJa33zxL+0Ko81hZkUZsU7XaKIOESB7gKnGzr47eHzbDpWy4FzLQDMSJvE12+ZyV3XZZKXOsnhCkUkHCjQw1R1Sxe/O1LDGweqOFjZCsB1U+N54vZZrCnMUIiLyMco0MNIVUsXr+w5x6ZjtRw/3wbAnCmT+cad13LXvExdwVBEPpECPQwcrW7l1eJKfrH7LP0eL0XTknhyTQG3zE5nuvbEReQKKdAd0tHTz4Z9lfxi9zmOn28jJjqKu+dN4W9um0mWrp0iIsOgQA+xyuZOXtxRwcu7z9Le3U9h1mT+4Z45/PG8KSRM0PVTRGT4FOghYK2luKKZH39wmneO1mCM4Y7CDB5dnsOC7ESdfi8iAaFAD6Lefi+/PlTNT7af4XBVK/Hjx/LnK/P4wrIctVVEJOAU6EHQcKGHX+w6y0s7K6hv72F66kS+dW8hn16QxYQYveUiEhxKlwCqbuniuS2neKX4HL39XlbNTOXRz+RyQ74uiCUiwadAD4CKxg6++95J3jxQDcBniqbyxRV55KdpyqGIhI4CfQQaL/Twow9O88P3TxMdZfj8smk8dkOe+uMi4ggF+jCca+rkB9tOsb64kt5+L/fNz2LdmgLSdXVDEXGQAv0qlNdf4F9/X8abB6uJMnD/wqk8sjyXmelxTpcmIqJAvxJ1bd08804Jr++rJHZMNI9cn8NjN+SREa89chEJHwr0T9Db7+UHW0/x3NZT9HssD1+fy5dXTyc1Ltbp0kREPkaBfhnWWn596DzPvFPC2aZObp+Tzro115KbMtHp0kREBqVAv0RFYwd//6vDbC9rZHbmZH7y8CJuLEhzuiwRkSEp0P36PF5eeL+c7757krHRUfzDPXP47JJpROuEIBFxCQU6sP9sM+teP0xJbTt3zMng6T+eowOeIuI6ozrQPV7Lc1vK+Jd3T5IaF8sLXyji1tnpTpclIjIsozbQmzt6+Zv1B9hcUs+nrsvk2/fNJX78WKfLEhEZtlEZ6DvLG/nb9Qepb+/hm/fM4XNLp+ma5CLieqMq0K21PLf1FM9uLCU7aQKvfmkZ865JcLosEZGAGDWB3tnbzxOvHuI3h89z59wM/um+64ifoBaLiESOEQW6MebrwGOABQ4Dj1hruwNRWCCdbexk7UvFlNa28+SaAtauzFOLRUQiTtRwf9AYkwX8FVBkrS0EooEHAlVYoJxv7eLBF3ZyvrWbnz6ymMdXTVeYi0hEGmnLZQww3hjTB0wAqkdeUuBUt3TxZy/spLWrj1+uXUphVrzTJYmIBM2w99CttVXA/wbOAueBVmvtxktfZ4xZa4wpNsYU19fXD7/Sq9TW3cdn/mMH9e09/OzRxQpzEYl4I2m5JAL3ALnAFGCiMeZzl77OWvu8tbbIWluUmpo6/EqvQp/Hy1d/sZ/atm5+9PAiFk5LDMl6RUScNOxAB24BTltr6621fcAG4PrAlDV8/R4vT7x6kK2l9Xzz3kKW5iU7XZKISEiMJNDPAkuNMROM7yjjzcDxwJQ1fN95p4T/PFDNE7fP4sHF2U6XIyISMiPpoe8CXgP24ZuyGAU8H6C6hqX4TBPPbyvns0uy+cqN+U6WIiISciOa5WKtfQp4KkC1jEhbdx9fX3+ArITx/P2d1zpdjohIyEXMmaLf+vUxqlu6Wf/4UibGRsywRESu2Eh66GHjg5MNrC+u5M9vyGPhtCSnyxERcYTrA72jp591Gw6RlzKRr90yw+lyREQc4/rexLd+c4yqli7WP76McWOjnS5HRMQxrt5DL6lp5+Xd53hsRS6LctRqEZHRzdWB/vLus8RER/GlVdOdLkVExHGuDfSuXg+v76tkzdwMkifFOl2OiIjjXBvoG4/V0N7dz58uusbpUkREwoJrA/2VPefIShjPklxdq0VEBFwa6NUtXewob+RPF11DdJRuViEiAi4N9I1Ha7AW7pyb6XQpIiJhw5WB/t6JOvLTJpGfNsnpUkREwobrAv1CTz+7ypu4cVZobpYhIuIWrgv0fRXN9Hq83DBDgS4iMpDrAn1baT0xY6IoytFt5UREBnJdoO883ciC7AQmxLj+MjQiIgHlqkBv6+7jWHWb5p6LiFyGqwJ975lmvBaW5OpCXCIil3JVoO863cTYaMP8bPXPRUQu5apAP1V/gbyUSYyP0XXPRUQu5apAr27pIjNhnNNliIiEJdcEusdrKau7wPRUnR0qInI5rgn0c02d9PR7mZUR53QpIiJhyTWBXt5wAYDpqRMdrkREJDy5J9DrOwDIS1HLRUTkclwT6LVt3YwbG0XChLFOlyIiEpZcE+h17T2kxY3DGN3QQkTkclwT6DWt3aRP1s2gRUQG45pAb+zoJTVOgS4iMhjXBHpLZx/x49U/FxEZjCsC3VpLW1cfkxXoIiKDckWgt3X30+vxkjJRLRcRkcG4ItBbO/sANGVRROQTuCLQL/T0AxA3TncpEhEZjCsCvaPXF+gTYxXoIiKDcUeg9yjQRUSG4opA7+n3AhA7xhXliog4whUJ6fFaAMZEuaJcERFHuCIh+/2BHu2KakVEnDGiiDTGJBhjXjPGnDDGHDfGLAtUYQN5vL6WS7T20EVEBjXSo4zfBX5nrb3fGBMDTAhATR/jz3OidKFFEZFBDTvQjTHxwErgYQBrbS/QG5iy/iuP9bVconTpXBGRQY2kh5EL1AM/McbsN8b80BjzsfvDGWPWGmOKjTHF9fX1w1qRtRd76Ap0EZHBjCTQxwALgOestfOBDmDdpS+y1j5vrS2y1halpqYOa0Wej1ouCnQRkcGMJNArgUpr7S7/96/hC/iA+6jlomOiIiKDGnZEWmtrgHPGmFn+RTcDxwJS1cfXBWgPXUTkk4x0lstXgZ/7Z7iUA4+MvKSPu3hiUbQCXURkUCMKdGvtAaAoQLUMyp/nROmgqIjIoFzRlfZ6L7ZcHC5ERCSMuSLQPZq2KCIyJFcEulcHRUVEhuSOQPcq0EVEhuKOQPcfFFXLRURkcK4IdI8OioqIDMkVge61FmPAqOUiIjIo1wS6TioSEflkrgh0j1cHREVEhuKKQLfW6sJcIiJDcEVMerxWe+giIkNwR6Crhy4iMiRXBLq1ujCXiMhQXBHovpaL01WIiIS3kV4PPSTmTJlMT7/H6TJERMKaKwL9gcXZPLA42+kyRETCmitaLiIiMjQFuohIhFCgi4hECAW6iEiEUKCLiEQIBbqISIRQoIuIRAgFuohIhDDW2tCtzJh6oGKYP54CNASwHLcYjeMejWOG0Tnu0ThmuPpxT7PWpg71opAG+kgYY4qttUVO1xFqo3Hco3HMMDrHPRrHDMEbt1ouIiIRQoEuIhIh3BTozztdgENG47hH45hhdI57NI4ZgjRu1/TQRUTkk7lpD11ERD6BAl1EJEK4ItCNMXcYY0qMMWXGmHVO1zNSxpgzxpjDxpgDxphi/7IkY8wmY8xJ/9dE/3JjjPmef+yHjDELBvw/D/lff9IY85BT4xmMMebHxpg6Y8yRAcsCNk5jzEL/+1jm/1nHb1Q4yJifNsZU+bf3AWPMnQOee9Jff4kx5vYByy/7mTfG5BpjdvmXv2KMiQnd6C7PGHONMWazMeaYMeaoMeav/csjfVsPNm7ntre1Nqz/AdHAKSAPiAEOArOdrmuEYzoDpFyy7DvAOv/jdcD/8j++E3gbMMBSYJd/eRJQ7v+a6H+c6PTYLhnTSmABcCQY4wR2+19r/D+7JkzH/DTw3y/z2tn+z3MskOv/nEd/0mceWA884H/8H8CXw2DMmcAC/+M4oNQ/tkjf1oON27Ht7YY99MVAmbW23FrbC/wSuMfhmoLhHuBn/sc/A+4dsPxF67MTSDDGZAK3A5ustU3W2mZgE3BHqIv+JNbabUDTJYsDMk7/c5OttTut79P+4oD/yzGDjHkw9wC/tNb2WGtPA2X4Pu+X/cz790pvAl7z//zA988x1trz1tp9/sftwHEgi8jf1oONezBB395uCPQs4NyA7yv55DfNDSyw0Riz1xiz1r8s3Vp73v+4Bkj3Px5s/G59XwI1ziz/40uXh6u/9LcXfnyx9cDVjzkZaLHW9l+yPGwYY3KA+cAuRtG2vmTc4ND2dkOgR6IV1toFwBrgK8aYlQOf9O+FRPx80tEyTuA5YDrwR8B54P84W05wGGMmAa8DX7PWtg18LpK39WXG7dj2dkOgVwHXDPh+qn+Za1lrq/xf64Bf4fuTq9b/pyX+r3X+lw82fre+L4EaZ5X/8aXLw461ttZa67HWeoEX8G1vuPoxN+JrT4y5ZLnjjDFj8YXaz621G/yLI35bX27cTm5vNwT6HmCG/2hvDPAA8KbDNQ2bMWaiMSbu4mPgNuAIvjFdPKr/EPCG//GbwBf8MwOWAq3+P2PfAW4zxiT6/6S7zb8s3AVknP7n2owxS/29xi8M+L/CysVQ87sP3/YG35gfMMbEGmNygRn4Dv5d9jPv38vdDNzv//mB759j/O//j4Dj1tpnBzwV0dt6sHE7ur2dPlJ8Jf/wHRUvxXck+BtO1zPCseThO4p9EDh6cTz4+mXvASeBd4Ek/3ID/Jt/7IeBogH/16P4DqyUAY84PbbLjPVlfH9y9uHr/30xkOMEivy/LKeA7+M/8zkMx/ySf0yH/L/UmQNe/w1//SUMmLkx2Gfe//nZ7X8vXgViw2DMK/C1Uw4BB/z/7hwF23qwcTu2vXXqv4hIhHBDy0VERK6AAl1EJEIo0EVEIoQCXUQkQijQRUQihAJdRCRCKNBFRCLE/wfRXW0FKR8R8wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "rankList = []\n",
    "for pep in validpeptides:\n",
    "    rankList.append(getRank(pep))\n",
    "    \n",
    "plt.plot(rankList)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'description': 'Asticcacaulis_benevestitus_DSM_16100_uid199072/NZ_AQWM01000023 -  gi|484226782|ref|NZ_AQWM01000023.1| Asticcacaulis benevestitus DSM 16100 = ATCC BAA-896 strain DSM 16100 B060DRAFT_scaffold_22.23_C, whole genome shotgun sequence - ORF 2',\n",
       " 'sequence': 'MAKPAPVACGIPPPAAAPSSSERTCSSKIQNLTERNSSW*',\n",
       " 'searchPattern': 'M[A-Z]{15,45}T[A-Z][A-Z]{6,8}[DE][A-Z]{5,30}\\\\*',\n",
       " 'searchRange': [3398, 3438],\n",
       " 'overallLength': 20838,\n",
       " 'rank': 12.220249405489058,\n",
       " 'closestB': {'strand': 4,\n",
       "  'motif': 'CLPRSIAL',\n",
       "  'start': 1397,\n",
       "  'end': 1404,\n",
       "  'score': 867.84,\n",
       "  'p-value': 8.38e-05,\n",
       "  'ORF': 2},\n",
       " 'closestC': {'strand': 1,\n",
       "  'motif': 'DALFTGQGGDNVFCQVPSAAPALD',\n",
       "  'start': 1685,\n",
       "  'end': 1708,\n",
       "  'score': -312.72,\n",
       "  'p-value': 8.95e-05,\n",
       "  'ORF': 3},\n",
       " 'ORF': 2,\n",
       " 'genome': 'Asticcacaulis_benevestitus_DSM_16100_uid199072'}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validpeptides[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some visualization\n",
    "Reference: https://matplotlib.org/examples/ticks_and_spines/tick-locators.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAi8AAAB8CAYAAAChUVA0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO2dd7heRbWH35VCUHoVkBJaqEoVQUW6IlchNKkKgdyLIIIFULCBgghSFQWlGKSFoiBFEAQiKBBqgNBCwAiRJiAhIUCAzP1jrZ0zZ5/9tUNyDlt+7/N8z/m+KWvPzF4zs2bN7H0spYQQQgghRF0Y0N8FEEIIIYToBBkvQgghhKgVMl6EEEIIUStkvAghhBCiVsh4EUIIIUStkPEihBBCiFoh40UIIYQQtULGixBCCCFqhYwXIYQQQtQKGS9CCCGEqBUyXoQQQghRK2S8CCGEEKJWyHgRQgghRK2Q8SKEEEKIWiHjRQghhBC1QsaLEEIIIWqFjBchhBBC1AoZL0IIIYSoFTJehBBCCFErZLwIIYQQolbIeBFCCCFErZDxIoQQQohaIeNFCCGEELVCxosQQgghaoWMFyGEEELUChkvQgghhKgVMl6EEEIIUStkvAghhBCiVsh4EUIIIUStkPEihBBCiFoh40UIIYQQtULGixBCCCFqhYwXIYQQQtQKGS9CCCGEqBUyXoQQQghRK2S8CCGEEKJWyHgRQgghRK2Q8SKEEEKIWiHjRQghhBC1QsaLEEIIIWqFjBchhBBC1AoZL0IIIYSoFTJehBBCCFErZLwIIYQQolbIeBFCCCFErZDxIoQQQohaIeNFCCGEELVCxosQQgghaoWMFyGEEELUChkvQgghhKgVMl6EEEIIUStkvAghhBCiVsh4EUIIIUStkPEihBBCiFoh40UIIYQQtULGixBCCCFqhYwXIYQQQtQKGS9CCCGEqBUyXoQQQghRK2S8CCGEEKJWyHgRQgghRK2Q8SKEEEKIWiHjRQghhBC1QsaLEEIIIWqFjBchhBBC1AoZL0IIIYSoFTJehBBCCFErZLwIIYQQolbIeBFCCCFErZDxIoQQQohaIeNFCCGEELVCxosQQgghaoWMFyGEEELUChkvQgghhKgVMl6EEEIIUStkvAghhBCiVsh4EUIIIUStkPEihBBCiFoh40UIIYQQ9SKl1KsPMApIwNkVccdF3NVZ2HzAUcB4YDrwMnAPcASwaJZuDHBaxXXKnzuyNJOAQyrKcQgwKfu9d0nGs8AlwPIlWUX868CjwKGAZWmGNihTAraOOjSKT6UyrQCcBfwTeBN4BrgZ2AuYq1SfzwI3Aq9G2e4HDgYGNLhHPwfeAf63Iq7cFs8DVwFrNLjPCXgLeBI4AZinlG448Je4r68DE4BzgY9l97WtNon0X4yyn9tEB78E3AG8Fp/bgd1KaVYqXeeVSLdNKd3ILM3MuA+jgeWALVuUPQF7hhwLWbfFfXoNeBg4DVilog5Lxn3/B5mOldKsA1yM6+ubwFPANdHmVkq7U+jPK3HtB4EfA4tlaebCdfpeYBowJXTpJ8DSkeZPwJ8blOcjUefNs7DNgKuBfwNvhJ5cWkpTtOOCDe5N/tkSOAz4DzB3RRkGAc8BP2pXb7LrrV0Rdx1wVvb7/Kwsb+P985fAAqV8kyPNRqXwo4FxDfSr6G9/BFYr5VscOIOu8eB5vF9tXkq3Oq6fLwAz6OqXC5bS/S275gzgCeAYsrEl2jIv2zRgHPDlkqxG/eCEivZ8DB8vlquIG53lfSvKdGxxn4FVs7glKtpnRsSvWVG+a4GX8DHoEeAUYJmI3zq77ky8f94PnAQsW5L1FeDFJmNPUYdfVsSdGnGXZWHzAcdHXd/A+8mtwE4Rfz1wTZP+n4BP4/3h4Io0lwB3AYOysDGNyv/f8Ol9Rp/UngKmkk1k0RGejc53dYQthBstk4F9gbXwSWELfII7LG9wehovNwBLlD4LZ2km0b7x8lrkXxIfcCfgA/zATNZRkWYoPui8BeyXyRkayvTZinLNBSyc/V490u6QhS0WctbHJ447gG2BYfgAuzNwE/DJ7JoH4IPy8cCawPLAfvgkdVFF3YfgnfgYYGxFfLkt1o9rTqL7wJa3/zLA7rjxeXqW5hh8gD8Z72DL4R3u20QHardNMpnX45PpNEoTRsSfiA8Ch0W7DQMOxwe2o7N0xYS1RVxnVXwSepNs4oj7PCVrj82BifggPlfpHp8L3FIKmxs3XEZH+3wP2BBYFvh41OXCinocAVyETz5bVcRvF2W9Cte3FYDVgBG48bFElva4uA8nAJ+kq4+NBr6b6cUtuJF5ELBelHFj3MA6MdLtiOvbMhVlOiXaxuL3VyPtqGi3obiBcyDwQJavkfFS3JtyP/oQ3vf2qCjDcHwCGtqu3tC58XJtlGVpfOJ7BjivlG8yPlHeWgqvMl5y/foY8Ne47/mEcxtwJz42LRfpDgV2ztJsGPX7Y6ZjX8AXWg/l9caNl9/EdZfFx5apwI9LY3bCx4QlgBWB70fYlhX3b1jpXs1Xqvun8XHkfOCoirYejetzUaYv4zp+csQXxstTwLcrxvR/UjJecF2eiS8EN8F1cGPg14SBQZfxskJce+Voj9txQ2ajTF47xstTeD+aOwsvxt1/0t14uQQ36LaJsq2L95v9In4XvO8uVXGt04DH4vse+PgyLIvfNcJWBT4KbB/hxdi7PqXF2n/Dp/cZfaC6Gh9AR2Th2+GryHPpMl5OxztbjxsT8blXYww9jZerW5RlEu0bL9NKafYIhV6lkSzcQ/T77PfQyLN+G+20aKTdtFxnfKC5m8aek2JyWBrv3KdUpBke8ncuhe8W5f4gPliVVylVbfGFkPWRZu0PnAk8G98/HnkOanVvW7VJFr9cdMYF8QH+K6X4T0T+Axvc8wSsG797TFi4MZ2A/bOwkcArJVl7RboVS+FnAH+puHahS5UDRbktQgcmAp/DDebRpfj58IHwkib6VehI0SaN7kNhMHwXHyQ/2kLeYHxV//1S/FzAi8ARWV+YARzfqs40Nl56GBNZnsuBGyvCrwKu71BvOjVeriilORV4vhQ2GTfm3gS2zcKrjJeyfm0f5VmtnX4RaQbgHoU7KY0b+OLideDULOxvlMYN3OgZm/0ujJfhpXRTgOMa3b8mZTwXX9BshU/i5XKOJpvYI+w84B/xvTBefgQ8Wkr3MN5XZhkvuDHyVhMdLPStMF7mLcUPBsZGuxb6347xclmUZ7csfBfciJxVR7yfTye8sw3kFf3q8FL43Li3JV/gX4obXANxI+wlot/jnqnfAhcA9+GG/LXAqs3uWR0/s+PMy9nAPtnvfaLx/K6ZDcAtw/NTSs9UCUjR6v3E6/F3cDnCnE3xle5bs/m6a+MeiBNSSjOrEmTtsjOu3MdXpLkCeBz3iOSMxNt8OvD7+N0QM1swk9Gqrq/T1V6744bpr1rUoRP2Aa5NKb2CD2rlsu+Br5R+XZH3tCjfblWCzWwu4H/jZzv1hArdaMDuwEMppT9VRVa0xaa4gXIDXs/hZrZIFr817rHqcd8rZBZt0ug+vJKV8bqU0gPN5KWU3gJ+B+xtZpYlGY4bB6Pi9854+1SWcTb07bOBzcxs+SLAzJbCDb6zS2lb6U2vMbMVcc9Xlc5Mwr15PzWzgW3KW4guHS1kvopPctuZ2dwNsq6HT+49xo2U0tP4pFmp+3HddYCNGtSjSDPQzHYD5m+WrkHeBfCty/PxLe5BwGfayJqPKQVXAIua2SdD9kb4ZH15Kd0ucZ2fVgnOdL+S0PVT8XZdo42y5lTNf+eU5Cd8IbCNmc3XoAwz8DYbUYraAZgXNwgLvoIvGg7DvWr3Ab8IOS+klEbgxu3awJIppc+llB7tsF7veWaH8XIhsL6ZrWxmS+AD7qgsfjF8sHssz2Rmt5nZtPhc2+IaW2dpi89xpTTHlNPg1n9DzGxp3CU7Gd8+6iYLX03djFvOP68QcUtFuRZoUZeCYfF3VruY2QIlWUdkaV9tZPzhK4ZVMjnL4y7TiyLod8CeZjaklG+euM5ruHW/K3BlM0U3sw3wCfDGrGxPppTeztIcUKrHsg1boaf8AbhX6LwIuhRYw8zWypINAybGoNONlNIbuOdvlVLUbXFPX8e3V57AV06NyrEM7sV5CveOtMMweur5iVk7lAfRffEtv7dTShNxL+aXSvKgu46sU2rbXSJqZbxN3qY5K1eU8dJM3v1Z1Fn4qnazUpmvzXRxGPBySunFTN7wUhk3alGm26xnP5on4q7Ft2vyQX0vfLt01iTWpt40vB7uVSjz+Yh/A9eBVXDdqeIYYCl6Tj45Rf9+Dd9u2Bn4Q9z7YgIbEfV4xcxuN7OfRZ8rKHTikQbXeBhYzMwWzsKK/vgmrmML41uLZS7Kxr0LcU9A2UAEmFxqv6WyuN2AR1JKj4RxdQFNjMhYIH4CP6t0Yyl6RuQvjIN9o1xvlNKtDLyQUnq50XXa4OH4u0KH+c4DNjazZWOc2xQfb8vsG3Evm9ndZvZzM9uslOYsYGUz+3Qp39UppeeLgJTSSxF+NL5Ft3exSDCzRc3sTGADfMv7OTO72sxW7rBe73netfGSUvoPPojsgw8qY1JKT7WRdRfcMrwc+ECLtLdE2vzzs1KakyrSnFQhK5+wn8Y9GjvEwFGWtQluvByVUrqtQtbuFdec2qIuzZiayXkmylbQagWbl39f3NX+XPweg6/ohpfyTI9rrYefn3k8/pbZOhvEb8fvx9ealOWCkLsnMA+d6dln8JXGnwBSSlOAK+l8FT2j9HsX/BzOdvhENKJiRZZPLk/hRusObRgEzTgWb4vv4vUCZnm6dqRrsiW+79tC3sN06chgular1jBHaw4Keafj9wuAMGL/TkweYdBtSfWElnNDyNsy5LXyRhRjQf6ZHmUoztLsFQYK+AR/fqnPdqI3Vder6t83R9wGuEfrStzD0oOYUI4DjjSzRuNZ0b/Xw8+wTQD2L8m5BDeCtsW3sjYG7jCzwxrIbETeNkV//ATuhT0jPLZlvhHpPoMfZD0wpfSPinQb073tns/iRtJTp7c1s8VKMoaHofQGPp7cENcvczbwRTP7EH7fqnTv3eh+WUZHnsKU0gu4zo2Iz3XZuJunu5Guw/+/xz08N5nZqVma8fj2VdHflscXDj3qHN7d+4BRKaXJWdSHgT+llPYApqSUDse32YaVZdSdQbNJzjm4W2sa8INS3L/xVdKqeWC4ODGzKfhebTOmF6uTJrxUTmNmL1XJwjvcTHz/+rUmsiaa2Y7A42Y2NqV0cynd5DbK1YjC07MqroTESmVilH1GKe0CZvbhlNK/KmStXsgwd1vvDSxlZvmkOwAfWC7OwlJW/kfNbEncW1NeEdwC/B/uQn6m5PGYgK88BhfhMXFMiQm6U0biK8Pp1rVbYSHv0PCsTMA9SYNLZcHc3b4Cvq+f83RK6XH8Xr4OXGZmq5VWa1Pxg3Qzgediy60TJtBTz18EXjSz50tp98T3s8dat10ZBprZx1NKY+muI3eFvDfp0pF8oC3aZFALY+vxijI+G/Kq+svZwC/DozgCd39fXbruwma2WErp3yHvNbzvTGtSjpynW/Sjc/CDzVuGAb0yvkrNaUdvGl7PzKru9WtZuq+a2a1RjqMblPMU/BDm1xvEzyz1t6Vww2KrPFFK6XX84PH1wFFmNgr4kZmdRJdOrI4/aFBmdfw8Wt72U4rrmtnuwCNmtmdK6fxS3ueycW8X4C4zuzf6Tc4/qrZiwsu1HrCOmZ2YRQ3ED+XmYTfgC6AZ+JhSqbMppQfM7DF8O2xiSuk+M1u1lGwCvr258Lvwvqwef5/sRd5ziG0bGt/7Ynvqr/E51syOBr5rZj8t+iDe3042s6/h/e0Z3PtYxdvxya9xP2545mF3dVadejC73vNyI66Ei+L7lLOICflifGBtZaT0BSmlNDGl9GQDw6Wc+D/4OYqTrTTLvEvG4a7fw6z1PvmlePseWo4ws+3xg4ijImhrYBH8hHm+Ovo8sIWZDW1ynZOBdc1sh1L49Gizf1Zs1VyEr66beWLaIlZn2+LnN/Kyr4UbFEW5LsD346u8RF/DT/xXuW6BWaugx/EngnJmZrrRqeEC3harm9m2baTdF5/syh6A6+jyvlyHb+cd3oa8C/E2OaAqMjMkL8I9aeu0IRP8KYm38XsyAn8EOR8wL434dsrYK1JKT+JekH3jMzZWqUBHevNuOQo43Hx7vKqcrwM/xJ+yW6QqTYkTgQ3b0JeHcQ/bEPwQ/mPAtzJPFDDLM7Yr3bfty2WcgXsDj7PG52pIKT2Ge5oabZNVMRJf6KxF9/twLD09iq9FX3uqDc/m2fiWSyOP3yX4027fqYpstYgys8H4KyceoWv7qBOuxb3kH8BfYdAuxbXmycJGx9/d8UXoqPA+dkxKadPe5KsLs8XzklJKZvZR/KT2mxVJjsC3YO4wsx/iq8hpuOtsU7q7HasYUjFgvFOs9PqAX+ED0s54RylYpKJcU2IQa0q02d74OxxuN7Nj8M4zEH/MdWm8Q5JSmmxm3wB+ER6Zc3EP0lb4QckzU9ch0ZH4mYR7S5ccHyuYfejpHSvK9KqZnYWv9q5IDQ4Sl/LcYWbHAz8zs+XwcyRP4Y+5/l8ka7fzfRk/CzC6fG0zuzzqdmFK6bZYhZ4QLvrCYN4Rf0LhO83O7QQnAhea2Qmp8VmiTrkA35a62MyOxVfOz+GPg+5BtIOZrYcP6nuklLoNlmZ2PnC6mX0jpTTVzEYCo83sGvzc1eP4YPdZfEIrdOTv0SYnxSR2OX6Wa0W83R7Gz2WcgBu4N5nZkfjTKC/j3owvULpXKaXXzGw07m1YiNIEklKaZGbfAk4xP2w8Cj9ztBA+kVKWWUE7/ehsfIX7Dj1Xt23pTYsytCSl9Bczm4iPZwc1SDYK+CY+8TzWIE0h7xUzOwf3qlyFPylyIV7PB3FP4Mfws1fXF4stMxuBey4uDz37F11b6eNp7BkqOB/XhQOo3lovOBG4x8zWrRhPuhGG0B543xtfivsNbvR9osH2eyvOxI3kyoO3KaUnzOxQ4ETzsz6/w59yWhr3cM7EPWIFi5vZvPg249rAt3Bv5GeKsyPBQDNbu3S5GeU+m1J6x8xWi++VhpiZ/Q3XjXvwBcma+Fj1IH7+rpA11cwuwQ2+BSkd/hUZqZePKdHiEeZyPL4qPBofRF+PzwP4o1yLZ+nG0N5L6iZnaSbRy0elK/I0kvWbKPsAmr+kbmQpX6vHglfCB+ancO/KFPzlRV8FhpTSfg5/F8vU7HrfyuKL92Ls3uBaP8LP+RSHG3u0BT7RzpLR6j5n+XaMsv0n8k/GVxEbV6Rt9Pj4w8CvGsj/DD4IrZiFfRnfI55O14untq9o3x6Px0YbPF5cj4pHWZvUtfJR6Uzufvi7e6bS9QKxs4hH0HFj+OEG+efHzwHsk4WtjxvNz0XbvgT8GZ8syo+h7oK7pafg7/EZj7+kLn8R5BDcGB8XbfcG/njnKZTemxLpN4g2/GuTNtkSX3W+GGV8Djcq/6eUZtajtjR/Sd3eJflz4wbKNHq+V6QtvWmkC5Gu5aPSmc69QdeLzyYDXy+lKV450PRR6QhfHvdcfTHqeCz++oT/xL15HDc4FyrlWyN04t9Rv4R71eYupevxqHSE/wDfApyHBo9KR7qb8DMUPe5fKd3uUY/FynERfydwTnzv8ah0KW3xqPSancTjBv11dL0o81Hcm1y8eLH8krqpuPFwMqWX6eFP9DScd9qoQ7f4aO/b8L5bPFRwBtXvdSlee9DjFQGldHdQ8YLA98uneKZd1AzzpzGuwQ2BzVLfeaHek8TK/WZ8Ffr51P2MgxD/1ZjZyfi23pYppbv7uzxCzGlkvNSYMGAOxl/k9If+Lk9/EwbMfriHYEw/F0eIPsXM9sO9aqelNrZ8hagzMl6EEEIIUSv0X6WFEEIIUStkvAghhBCiVnT8qLRts39i/TNhwDuQLDEg/Sr9MB04JwonhBBCCFGm4zMvdqSlbi9jToDxSxkwQgghhOgLOt82Kr9j1n9/ZTaURQghhBCiJZ17Xha2RNVLpZ/lntlTpLZYBH/ZT3+xFP4/J/qL/q4/qA1AbdDf9Qe1AagN+rv+oDYAf3nhSn11sc6NF7OUUpqd/+OnY8xsTOrH/9vQ323Q3/WPMqgN3udt0N/1jzKoDd7nbdDf9Y8yqA36uA30tJEQQgghakVdjZdR/V2AfmZUfxfgPcCo/i7Ae4BR/V2A9wCj+rsA7wFG9XcB+plR/V2A9wCj+rsAfU0tt436G7WB2gDUBu/3+oPaANQGoDaAemwbHTXbS1E/1AZqA1AbvN/rD2oDUBuA2gD6uA30v42EEEIIUSvqeuZFCCGEEO9TZLwIIYQQolbIeBFCCCFErejIeDGzrc3sMTObaGbfmVOFmlOY2dxmdqeZ3W9mD5nZURF+QdRrvJmdY2aDI3whM7vczB6IfGs2k9MbWf2BmS1oZpeZ2aNm9oiZbWRmR5rZv8xsXHy2ibSDzexcM3sw0h7eTE6Edyyrj+t/cNyfh8zs6xG2tpndEeW928w2iPBDs3qMN7N3zGzhRnJ6K6sP6nyOmb1gZuOzsB+HPo4zs+vNbKkIX9XMbjezN83skCz9Kln5x5nZq1n7dSQr4ir1py/bIMK/FmV4yMyOL8Uta2bTSu3QsNydyorwgWZ2n5ldPXtr3JMGerBW3KMHzewqM5s/wvco3e+ZZrZ2xI0xH+eKuMUjfDkzuzF0YYyZLR3hm5VkvWFmwyPubPOx9IFo13nnYP2XMbObzezhuEcHR3gj/TUz+7n5nPeAma2byXonq8+VWXhlfcxsiJldHLLGmtnQCN/KzO6J9r/HzDafU/Vv1gYR10N/zWyDrJ73m9n2JXk99Dfa7RgzmxB95KAI3y5r57vN7FNZnuvM7JW2+0FKqa0PMBB4AlgBmAu4H1i93fzvhQ/+n5jmje+DgbHAhsA2EWfARcD+keZnwA/j+6rAjc3kxO+OZPVTO5wLjIzvcwELAkcCh1Sk3R0YHd8/CEwChjaSE987ltWHdV8TGB/XHwT8BVgJuB74XHYPx1Tk/QJwUzM5EdeRrD6q96eBdYHxWdj82feDgDPi++LAx4Bjqu5jpBkIPAcs11tZjfSnj9tgs7h3Q4rylvJcBlyal72J3ncsK8K/CVwIXN1PenAXsEl83wf4cUW+jwBPZL/HAOtXpLsU2Cu+bw6cV5FmYeBl4IMVunMS8J05WP8lgXXj+3zABGD1Jvq7DXAtPp5vCIzN0k1rcI3K+gAHZHJ3BS6O7+sAS8X3NYF/zWEdaNQGlfpLjHFZ3heK3430FxgB/A4YUJI1L10PCn0UeDTLswU+LrbVDzrxvGwATEwpPZlSmgGMBrbrIH+/k5xp8XNwfFJK6U8Rl4A7gaUjzerATZH3UWComX2okZxI15GsOVnfKsxsAXwAOzvKMiOl9EqTLAmYx8wGAR8AZgCv9kJOQ1nvpj69YDV8AJqeUnob+CuwQ5Rt/kizANX/p2Q33CBtJodeyJrjpJRuwSeMPCxv+3no0uEXUkp3AW81EbkFPpn9szeyeqk/74qqNgD2B36aUnqzKG9WxuHAP4CH2ix3R7IifGngf4CzZkMVW9KgDYYBt8T3G4AdK7Luho/5rZg1zgE3Uz1H7ARcm1KaHmV6FXy1jo8Lc+wR2JTSsymle+P7VOAR4MON9Bcv/+9iSL8DWNDMlmxxjUb12Q43fMEN2S3MzFJK96WUijHiIeADZjbkXVW0efkq24AG+puNcQBzZ/Vppr/7Az9KKc0syZoWcyN0b2dSSjcCU9utRyfGy4eBp7PfkyOsVoSLaxxuPd6QUhqbxQ0GvgRcF0H3ExOSuet/OcIYaSanU1l9zPLAv4HfhqvvLDObJ+IODJfeOWa2UIRdBrwGPAs8BZyQUnq5hZxOZfUl44GNzWwRM/sgvrJaBvg68DMzexo4Aei2pRVptwZ+30IOvZDVb4Rr92lgD+AHHWTdlZLx1aGsVvrTVwzD7+NYM/urmX0MIFz936bnuyualbtTWQCnAIcBM2d7zdrnIbqMjJ3p0uOcXehpbP823P/fj4kasnEO2B6Yz8wWKeWr0p3f4p68VYFf9KoWHRLbNuvgnvNG+tts3ps7tj7uCOM0l11Vn1mywhiYgv9DxZwdgXsLA2JOU2qDSv2NdB83s4eAB4GvZMZMI/1dEdgl2udaM1s5k7W9mT0KXIN7+nrF++7AbkrpnZTS2rjhsIF1P3vyK+CWlNKt8funuKU9DvgacB/wThtyOpLVxwzC3canp5TWwY2J7wCn4wq3Nm5cnBjpN4hyLoUP3N8ysxWayKEXsvqMlNIjwHH41s51wLgo0/7AN1JKywDfIFbWGV8A/l4YW03k0Kms/iSl9N0o5wXAge3kMbO5gG3xLYLeymqmP33JIHwbY0PgUOCSmIiPBE7OPKx5+kbl7kiWmX0eeCGldM8cqFcn7AMcYGb34NsIM/JIM/s4MD2llJ8V2iOl9BFg4/h8KcIPATYxs/uATYB/kY1z4bX4CPDn/BoppRH4uPAIbijNUcKg/D3w9cJT0ou+sFxKaX18O/wUM1uxiOhNfcxsDXxM2a+TuvSWijZopL+klMamlNbAt4APNz/32Ux/hwBvRPucCZxTRKSULk8prQoMB37c6wqk9vfJNgL+nP0+HDi83fzvxQ9uXR8S338IXEHs0VWkNfyMxvzN5LxbWX1Q5yWASdnvjYFrSmmGEnviwC+BL2Vx5wBfbEdOu7L6WQd+gu9FT6FrL9aAV0vpLgd2byUnvr8rWXOwrrPuRUXcsuU4Gp9d2g64vsl1WspqV3/mdBvghudm2e8ngMWAW6OPTgJewbdaDmxW7l7IOhZfyU/CV+nTgfP7WQ+GAXeWwk4Gjmgib2/gtIrweYHJpbCDgd80kfVp5vDZH3yb/8/AN1vpL/BrYLcs7jFgyYo8o4CdmtUnrrlRfB8EvJiNE0vjZ08+Oafvf6M2aKS/FXlvAtZvpr/Ao8Dy8d2AKQ3K8SSwaPZ703bvfyeel7uAlc1s+Vh57Qpc2SLPewozW8zMFozvHwC2Ah41s5HAZ3ElnZmlXzDqCjAS9xHqhD0AAALQSURBVKS82khO/O5I1pysbxUppeeAp81slQjaAni4tI+7Pb4tAr69szlAuMc3xA9ZVcqJdB3Jmn21aw/rejJiWdzFfSF+LmWTSLI58HiWfoGI+2MbcuiNrP4gd+XiBkm796LHeZ1OZTXTnz7mCvygImY2DD+A+2JKaeOU0tCU0lDcNf6TlNJpLcrdqazDU0pLR/iu+AHuPfugzt3I9HgA8D3gjCxuAL5YGZ2FDTKzReP7YODzRB83s0UjD/gCd9aKO+imO+asVHzHPXpzbEyIa5wNPJJSOikLb6S/VwJfjnJuiE/Cz5o/PTok8i4KfBIfR5vV50pgr/i+E36/U8wl1+AHe/8+B6rdjUZtQAP9jTl/UIQvh2+FTWqhv7Nk4ePdhMi/UuHNMX9yawjwUq8q0qG1tk0U4gngu31hIc5ma/Oj+HbNA3hn+0GEvx11GhefInyjqO9jwB+AhZrJ6Y2sfmqHtYG7o/xXAAsB5+H7mQ/gnWzJSDsvvj3wED5IH9pMToR3LKuP639rXP9+YIsI+xRwT4SNBdbL0u9NPCXVSk5vZfVBnS/Ct/DewldL++Iu4/Fxn67CDy6Cexcm44epX4nv80fcPPhgs0BJfm9kVepPH7fBXMD5UfZ7gc0r8h1Jd69RI73vWFYWvil987RRVRscjI9NE/DtbSuV646SjHlCvx+IvnwqMDDidsKN9Qn4Ic4hWb6h+DbSgCxsAPB3fLwYj2/ZzDGPdPTNFGUvxuhtmuiv4R7jJ6KM60f4J+L3/fF331b1wQ+7XgpMxB/mWCHCv4dvP47LPov3QxtU6i++JfhQpLsXGN5Kf/EnWK+JdrgdWCvCv53Juh34VJbnVvw82euhm59tVg/9byMhhBBC1Ir33YFdIYQQQtQbGS9CCCGEqBUyXoQQQghRK2S8CCGEEKJWyHgRQgghRK2Q8SKEEEKIWiHjRQghhBC14v8BdpsDzW6az14AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 576x432 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "\n",
    "peptide = validpeptides[len(validpeptides) - 1]\n",
    "length = peptide[\"overallLength\"]\n",
    "\n",
    "# Setup a plot such that only the bottom spine is shown\n",
    "def setup(ax):\n",
    "    ax.spines['right'].set_color('none')\n",
    "    ax.spines['left'].set_color('none')\n",
    "    ax.yaxis.set_major_locator(ticker.NullLocator())\n",
    "    ax.spines['top'].set_color('none')\n",
    "    ax.xaxis.set_ticks_position('bottom')\n",
    "    ax.tick_params(which='major', width=1.00)\n",
    "    ax.tick_params(which='major', length=5)\n",
    "    ax.tick_params(which='minor', width=0.75)\n",
    "    ax.tick_params(which='minor', length=2.5)\n",
    "    ax.set_xlim(0, length)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.patch.set_alpha(0.0)\n",
    "    \n",
    "plt.figure(figsize=(8, 6))\n",
    "n = 8\n",
    "\n",
    "    \n",
    "# Multiple Locator\n",
    "ax = plt.subplot(n, 1, 2)\n",
    "setup(ax)\n",
    "ax.xaxis.set_major_locator(ticker.MultipleLocator(length / 8))\n",
    "ax.xaxis.set_minor_locator(ticker.MultipleLocator(length / 16))\n",
    "ax.text(0.0, 0.1, peptide[\"sequence\"], fontsize=14,\n",
    "        transform=ax.transAxes)\n",
    "\n",
    "# Push the top of the top axes outside the figure because we only show the\n",
    "# bottom spine.\n",
    "plt.subplots_adjust(left=0.05, right=0.95, bottom=4, top=6)\n",
    "\n",
    "plt.scatter(peptide[\"searchRange\"][0], 0, c=\"red\")\n",
    "plt.scatter(peptide[\"searchRange\"][1], 0, c=\"red\")\n",
    "plt.scatter(peptide[\"closestB\"][\"start\"], 0, c=\"blue\")\n",
    "plt.scatter(peptide[\"closestB\"][\"end\"], 0, c=\"blue\")\n",
    "plt.scatter(peptide[\"closestC\"][\"start\"], 0, c=\"green\")\n",
    "plt.scatter(peptide[\"closestC\"][\"end\"], 0, c=\"green\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
