<html>
    <head>
        <title>Lasso Miner</title>
        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
        <link href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh" crossorigin="anonymous">
        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js" integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6" crossorigin="anonymous"></script>
        <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
        <script type=”text/javascript” src=”http://www.ncbi.nlm.nih.gov/projects/sviewer/js/sviewer.js”></script>
    </head>
    <body>
            <nav class="navbar navbar-expand-lg navbar-light bg-light">
                    <a class="navbar-brand" href="#">Lasso Miner</a>
                    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
                        <span class="navbar-toggler-icon"></span>
                    </button>
                    <div class="collapse navbar-collapse" id="navbarNav">
                        <ul class="navbar-nav">
                        <li class="nav-item">
                            <a class="nav-link" href="$BASE_URL$/status.html">Status</a>
                        </li>
                        <li class="nav-item">
                            <a class="nav-link" href="$BASE_URL$/matches.html">Matches</a>
                        </li>
                        <li class="nav-item active">
                            <a class="nav-link" href="$BASE_URL$/about.html">About <span class="sr-only">(current)</span></a>
                        </li>
                        </ul>
                    </div>
                </nav>
        <div id="canvas">
            <div class="container">
                <h2 id="lasso-peptide-mining-software">Lasso Peptide Mining Software</h2>
<p><em>Author: Nathan Alam - Advised by James Link and Larry Cho</em></p>
<p>This is a project created as a result of independent research with Princeton University&#39;s Chemical and Biological Engineering department. </p>
<p>This web tool is intended to launch and view results from requests to mine genomes for lasso peptides. To use the tool, navigate to the <strong>Status</strong> section in the navigation pane to access the <strong>Launch</strong> tool, which allows you to launch a search for lasso peptides in a specified genome. From that page, you can also monitor the progress of the search, with the ability to view the progress on the run, and a histogram upon clicking the <em>Details</em> button corresponding to the rank distribution of the lasso peptides found so far. The results, as they are entered into the database, can be viewed in the <strong>Matches</strong> site, also accessible from the top navigation pane.</p>
<h4 id="launching-a-run">Launching a run</h4>
<p>First, navigate to the <a href="$BASE_URL$/status.html">Status page</a>, and then your options will depend on the state of the server and the runs at the time - there may be no active runs, and there may be no stored runs at all, or there may be several runs and one ongoing. As of now, the server can only handle running <em>one run at a time</em>, and so you will have to either wait or cancel a run before attempting a new one.</p>
<p>Note that, while initiating or deleting a run, the server may be too occupied to handle requests, which will result in a display &quot;Server is busy, refresh to try again&quot; - if this occurs, you may have to wait a minute before refreshing the page.</p>
<h5 id="observing-a-past-run">Observing a past run</h5>
<p>Upon loading the page, a list of cards corresponding to past runs should be available. Hit the <strong>Details</strong> button for more information on the run, such as which accessions it ran on, and a histogram on the results so far, and hit the <strong>Delete</strong>  button to remove the run and all of the lasso peptides associated with it.</p>
<h5 id="launching-a-new-run">Launching a new run</h5>
<p>Hit the <strong>Launch</strong> button to view a list of queries to fill out prior to a run. For the <em>Accessions</em> field, you must enter a comma separated list of all of the genomes that you wish to run on. The <em>Accessions</em> themselves are simply queries that the server will automatically launch to NCBI and retrieve genomes from. Note that very generic search times may result in excessively long wait times (ex: Typing in &quot;Bacteria&quot; would result in millions of records that would need to be downloaded, which would most likely crash the run). </p>
<p>Instead, it is recommended to use <strong>GI Numbers</strong> to refer to a genome uniquely. Tools for generating GI numbers for large sets of genomes can be found on the <a href="https://github.com/NCBI-Hackathons/EDirectCookbook">EDirectCookbook github page</a></p>
<p>Specifically the code that gets run for each accession is as follows:</p>
<pre><code class="lang-python"><span class="hljs-keyword">if</span>(<span class="hljs-string">".gz"</span> <span class="hljs-keyword">in</span> accession[len(accession) - 3:]):
        <span class="hljs-keyword">print</span>(<span class="hljs-string">"found a direct zip file address"</span>)
        os.system('echo <span class="hljs-string">"' + accession +'"</span> &gt;&gt; fileurls.txt')
    <span class="hljs-keyword">else</span>: 
        <span class="hljs-keyword">print</span>(<span class="hljs-string">"attempting assembly search"</span>)
        str1 = 'esearch -<span class="hljs-keyword">db</span> assembly -<span class="hljs-keyword">query</span> <span class="hljs-string">"' + accession + '"</span> | efetch -<span class="hljs-keyword">format</span> docsum | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F<span class="hljs-string">"/"</span> \'{<span class="hljs-keyword">print</span> <span class="hljs-variable">$0</span><span class="hljs-string">"/"</span><span class="hljs-variable">$NF</span><span class="hljs-string">"_genomic.fna.gz"</span>}\' &gt;&gt; fileurls.txt'
        os.system(str1)
        <span class="hljs-keyword">print</span>(<span class="hljs-string">"attempting nucleotide search"</span>)
        str1 = 'esearch -<span class="hljs-keyword">db</span> nucleotide -<span class="hljs-keyword">query</span> <span class="hljs-string">"' + accession + '"</span> | efetch -<span class="hljs-keyword">format</span> fasta &gt;&gt; genomes/' + accession + '.fna'
        os.system(str1)
</code></pre>
<h5 id="cancelling-a-run">Cancelling a run</h5>
<p>You can also cancel a run in progress by hitting the cancel button once the progress bar appears. This will not delete the peptides associated with the run, or the record of the run, but will just halt the run where it is and allow the user to put in another run (as of right now, it is not possible to resume the run at a later point).</p>
<h4 id="viewing-results">Viewing Results</h4>
<p>Runs can be viewed on the <a href="$BASE_URL$/matches.html">Matches page</a>, which will begin with an empty page and a <strong>Search</strong> button. Upon hitting the button, a series of queries will pop up allowing you to query the database with certain specifications. </p>
<h5 id="search-queries">Search Queries</h5>
<p>The only non-optional specification is the <em>Max Peptides</em> option, which you should set to a reasonable number for the server to retrieve. The number N entered here will result in a query of the top N ranked peptides in the set.</p>
<p>Adding other parameters to the search query will narrow down the list of peptides to match those parameters. For example, specifying the genome will only permit peptides corresponding to that genome, and specifying a run name will only permit peptides associated with that run.</p>
<h5 id="interacting-with-results">Interacting with results</h5>
<p>Each peptide result will generate a card, with a set of properties immediately available, the predicted sequence in the header of the card, and a link to the NCBI page that would be associated with that peptide (clicking the link will lead you to the region of the genome published on NCBI that corresponds to the peptide). Also on each card is an embedded view that contains the NCBI annotated genome sequence, with the initial zoom set to where the peptide is expected to be.</p>
<p>Hitting the <strong>Details</strong> button will expand the card to show a generated plot that lacks detailed sequence data, but does indicate the location of nearby motif matches, their motif sequences, and the p-value associated with the motif match. For each of these plots, the green curve is the peptide itself, and curves associated with other motifs are labelled accordingly. This can be helpful even if the genome on NCBI has not yet been annotated to give a better sense of how realistic the peptide prediction actually is.</p>
<h4 id="additional-resources">Additional Resources</h4>
<ul>
<li><p>Fork the <a href="https://github.com/nathanalam/lassomining.git">repo</a>!</p>
<pre><code class="lang-bash">  git <span class="hljs-keyword">clone</span> <span class="hljs-title">https</span>://github.com/nathanalam/lassomining.git
  cd lassomining
</code></pre>
</li>
<li><p>Read the <a href="https://drive.google.com/open?id=1RNZdafP2ZnjEFHseWwcHnbOavtkWycgC">associated paper</a></p>
</li>
<li><p>Contact the <a href="nathanalam99@gmail.com">author</a></p>
</li>
</ul>

            </div>
        </div>
    </body>
</html>